<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Inferring the underlying source-specific 3D tensor — tensor • Unico</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Inferring the underlying source-specific 3D tensor — tensor"><meta property="og:description" content="Infers the underlying (sources by features by observations) 3D tensor from the observed (features by observations) 2D mixture, under the assumption of the Unico model that each observation is a mixture of unique source-specific values (in each feature in the data). In the context of bulk genomics containing a mixture of cell types (i.e. the input could be CpG sites by individuals for DNA methylation and genes by individuals for RNA expression), tensor allows to estimate the cell-type-specific levels for each individual in each CpG site/gene (i.e. a tensor of CpG sites/genes by individuals by cell types)."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Unico</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Unico-Tutorial.html">Unico-Tutorial</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/cozygene/Unico/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Inferring the underlying source-specific 3D tensor</h1>
    <small class="dont-index">Source: <a href="https://github.com/cozygene/Unico/blob/HEAD/R/Unico.r" class="external-link"><code>R/Unico.r</code></a></small>
    <div class="hidden name"><code>tensor.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Infers the underlying (sources by features by observations) 3D tensor from the observed (features by observations) 2D mixture, under the assumption of the Unico model that each observation is a mixture of unique source-specific values (in each feature in the data). In the context of bulk genomics containing a mixture of cell types (i.e. the input could be CpG sites by individuals for DNA methylation and genes by individuals for RNA expression), <code>tensor</code> allows to estimate the cell-type-specific levels for each individual in each CpG site/gene (i.e. a tensor of CpG sites/genes by individuals by cell types).</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">tensor</span><span class="op">(</span></span>
<span>  <span class="va">X</span>,</span>
<span>  <span class="va">W</span>,</span>
<span>  <span class="va">C1</span>,</span>
<span>  <span class="va">C2</span>,</span>
<span>  <span class="va">Unico.mdl</span>,</span>
<span>  parallel <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  num_cores <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  log_file <span class="op">=</span> <span class="st">"Unico.log"</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  debug <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>X</dt>
<dd><p>An <code>m</code> by <code>n</code> matrix of measurements of <code>m</code> features for <code>n</code> observations. Each column in <code>X</code> is assumed to be a mixture of <code>k</code> sources. Note that <code>X</code> must include row names and column names and that NA values are currently not supported. <code>X</code> should not include features that are constant across all observations. Note that <code>X</code> could potentially be different from the <code>X</code> used to learn <code>Unico.mdl</code> (i.e. the original observed 2D mixture used to fit the model).</p></dd>


<dt>W</dt>
<dd><p>An <code>n</code> by <code>k</code> matrix of weights - the weights of <code>k</code> sources for each of the <code>n</code> mixtures (observations). All the weights must be positive and each row - corresponding to the weights of a single observation - must sum up to 1. Note that <code>W</code> must include row names and column names and that NA values are currently not supported.</p></dd>


<dt>C1</dt>
<dd><p>An <code>n</code> by <code>p1</code> design matrix of covariates that may affect the hidden source-specific values (possibly a different effect size in each source). Note that <code>C1</code> must include row names and column names and should not include an intercept term. NA values are currently not supported. Note that all covariates in <code>C1</code> must be present and match the order of the set of covariates in <code>C1</code> stored in <code>Unico.mdl</code> (i.e. the original set of source-specific covariates available when initially fitting the model).</p></dd>


<dt>C2</dt>
<dd><p>An <code>n</code> by <code>p2</code> design matrix of covariates that may affect the mixture (i.e. rather than directly the sources of the mixture; for example, variables that capture biases in the collection of the measurements). Note that <code>C2</code> must include row names and column names and should not include an intercept term. NA values are currently not supported. Note that all covariates in <code>C2</code> must be present and match the order of the set of covariates in <code>C2</code> stored in <code>Unico.mdl</code> (i.e. the original set of not source-specific covariates available when initially fitting the model).</p></dd>


<dt>Unico.mdl</dt>
<dd><p>The entire set of model parameters estimated by Unico on the 2D mixture matrix (i.e. the list returned by applying function <code>Unico</code> to <code>X</code>).</p></dd>


<dt>parallel</dt>
<dd><p>A logical value indicating whether to use parallel computing (possible when using a multi-core machine).</p></dd>


<dt>num_cores</dt>
<dd><p>A numeric value indicating the number of cores to use (activated only if <code>parallel == TRUE</code>). If <code>num_cores == NULL</code> then all available cores except for one will be used.</p></dd>


<dt>log_file</dt>
<dd><p>A path to an output log file. Note that if the file <code>log_file</code> already exists then logs will be appended to the end of the file. Set <code>log_file</code> to <code>NULL</code> to prevent output from being saved into a file; note that if <code>verbose == FALSE</code> then no output file will be generated regardless of the value of <code>log_file</code>.</p></dd>


<dt>verbose</dt>
<dd><p>A logical value indicating whether to print logs.</p></dd>


<dt>debug</dt>
<dd><p>A logical value indicating whether to set the logger to a more detailed debug level; set <code>debug</code> to <code>TRUE</code> before reporting issues.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A <code>k</code> by <code>m</code> by <code>n</code> array with the estimated source-specific values. The first axis/dimension in the array corresponds to the different sources.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>After obtaining all the estimated parameters in the Unico model (by calling <a href="Unico.html">Unico</a>), <code>tensor</code> uses the conditional distribution \(Z_{jh}^i|X_{ij}=x_{ij}\) for estimating the \(k\) source-specific levels of each sample \(i\) at each feature \(j\).</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">data</span> <span class="op">=</span> <span class="fu"><a href="simulate_data.html">simulate_data</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">100</span>, m <span class="op">=</span> <span class="fl">2</span>, k <span class="op">=</span> <span class="fl">3</span>, p1 <span class="op">=</span> <span class="fl">1</span>, p2 <span class="op">=</span> <span class="fl">1</span>, taus_std <span class="op">=</span> <span class="fl">0</span>, log_file <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> INFO [2024-01-11 06:43:04] Start simulation ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> INFO [2024-01-11 06:43:04] Finished simulation</span>
<span class="r-in"><span><span class="va">res</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">res</span><span class="op">$</span><span class="va">params.hat</span> <span class="op">=</span> <span class="fu"><a href="Unico.html">Unico</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">X</span>, <span class="va">data</span><span class="op">$</span><span class="va">W</span>, <span class="va">data</span><span class="op">$</span><span class="va">C1</span>, <span class="va">data</span><span class="op">$</span><span class="va">C2</span>, parallel<span class="op">=</span><span class="cn">FALSE</span>, log_file<span class="op">=</span><span class="cn">NULL</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> INFO [2024-01-11 06:43:04] Starting Unico ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> INFO [2024-01-11 06:43:04] Validate Unico inputs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> INFO [2024-01-11 06:43:04] Starting parameter learning ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   |                                                  | 0 % ~calculating    |+++++++++++++++++++++++++                         | 50% ~00s            |++++++++++++++++++++++++++++++++++++++++++++++++++| 100% elapsed=00s  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> INFO [2024-01-11 06:43:04] Formating results ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> INFO [2024-01-11 06:43:04] Capping extreme values in estimated parameters</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> INFO [2024-01-11 06:43:04] Finished parameter learning</span>
<span class="r-in"><span><span class="va">res</span><span class="op">$</span><span class="va">Z</span> <span class="op">=</span> <span class="fu">tensor</span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">X</span>, <span class="va">data</span><span class="op">$</span><span class="va">W</span>, <span class="va">data</span><span class="op">$</span><span class="va">C1</span>, <span class="va">data</span><span class="op">$</span><span class="va">C2</span>, <span class="va">res</span><span class="op">$</span><span class="va">params.hat</span>, parallel<span class="op">=</span><span class="cn">FALSE</span>, log_file<span class="op">=</span><span class="cn">NULL</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> INFO [2024-01-11 06:43:04] Validate tensor inputs ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> INFO [2024-01-11 06:43:04] Starting tensor ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   |                                                  | 0 % ~calculating    |+++++++++++++++++++++++++                         | 50% ~00s            |++++++++++++++++++++++++++++++++++++++++++++++++++| 100% elapsed=00s  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> INFO [2024-01-11 06:43:04] Formating tensor result ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> INFO [2024-01-11 06:43:04] Finished tensor estimation</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Zeyuan Chen, Elior Rahmani.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

