{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: Biobase\n",
      "\n",
      "Warning message:\n",
      "“package ‘Biobase’ was built under R version 4.1.2”\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Warning message:\n",
      "“package ‘BiocGenerics’ was built under R version 4.1.1”\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Setting options('download.file.method.GEOquery'='auto')\n",
      "\n",
      "Setting options('GEOquery.inmemory.gpl'=FALSE)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(\"GEOquery\")\n",
    "set.seed(2023)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "data.dir = \"../Data/Methylation/Purified-Reinius/\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Full Hannum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Found 1 file(s)\n",
      "\n",
      "GSE40279_series_matrix.txt.gz\n",
      "\n",
      "Using locally cached version: ../Data/Methylation/Purified-Reinius//GSE40279_series_matrix.txt.gz\n",
      "\n",
      "Warning message in data.table::fread(text = dat[(series_table_begin_line + 1):(series_table_end_line - :\n",
      "“Stopped early on line 324059. Expected 657 fields but found 0. Consider fill=TRUE and comment.char=. First discarded non-empty line: <<\"cg18281842\"\t0.6641291\t0.6416066\t0.7031253\t0.6512614\t0.6589804\t0.6285009\t0.6797004\t0.6575938\t0.6719682\t0.6271186\t0.6081633\t0.6271322\t0.6447954\t0.671915\t0.6212175\t0.6527885\t0.6725338\t0.6455413\t0.6605631\t0.5844895\t0.6939166\t0.6064965\t0.6043743\t0.6334873\t0.648257\t0.5848833\t0.5829015\t0.567418\t0.5912762\t0.6026103\t0.6069624\t0.5357862\t0.6403599\t0.6880816\t0.7044414\t0.6404482\t0.5954085\t0.63186\t0.7182326\t0.6465346\t0.632231\t0.5789111\t0.6757559\t0.653791\t0.6175523\t0.7231798\t0.6113838\t0.74055\t0.6571084\t0.6306>>”\n",
      "Using locally cached version of GPL13534 found here:\n",
      "../Data/Methylation/Purified-Reinius//GPL13534.soft.gz \n",
      "\n"
     ]
    }
   ],
   "source": [
    "Sys.setenv(\"VROOM_CONNECTION_SIZE\" = 131072 * 10)\n",
    "gse <- GEOquery::getGEO(\"GSE40279\", destdir = data.dir, GSEMatrix = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "List of 1\n",
      " $ GSE40279_series_matrix.txt.gz:Formal class 'ExpressionSet' [package \"Biobase\"] with 7 slots\n",
      "  .. ..@ experimentData   :Formal class 'MIAME' [package \"Biobase\"] with 13 slots\n",
      "  .. .. .. ..@ name             : chr \"Justin,,Guinney\"\n",
      "  .. .. .. ..@ lab              : chr \"\"\n",
      "  .. .. .. ..@ contact          : chr \"justin.guinney@sagebase.org\"\n",
      "  .. .. .. ..@ title            : chr \"Genome-wide Methylation Profiles Reveal Quantitative Views of Human Aging Rates\"\n",
      "  .. .. .. ..@ abstract         : chr \"Genome wide DNA methylation profiling of individuals across a large age range. The Illumina Infinium 450k Human\"| __truncated__\n",
      "  .. .. .. ..@ url              : chr \"https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE40279\"\n",
      "  .. .. .. ..@ pubMedIds        : chr \"23177740\"\n",
      "  .. .. .. ..@ samples          : list()\n",
      "  .. .. .. ..@ hybridizations   : list()\n",
      "  .. .. .. ..@ normControls     : list()\n",
      "  .. .. .. ..@ preprocessing    : list()\n",
      "  .. .. .. ..@ other            :List of 24\n",
      "  .. .. .. .. ..$ contact_address        : chr \"2606 E Roy St\"\n",
      "  .. .. .. .. ..$ contact_city           : chr \"Seattle\"\n",
      "  .. .. .. .. ..$ contact_country        : chr \"USA\"\n",
      "  .. .. .. .. ..$ contact_email          : chr \"justin.guinney@sagebase.org\"\n",
      "  .. .. .. .. ..$ contact_institute      : chr \"Sage Bionetworks\"\n",
      "  .. .. .. .. ..$ contact_name           : chr \"Justin,,Guinney\"\n",
      "  .. .. .. .. ..$ contact_state          : chr \"WA\"\n",
      "  .. .. .. .. ..$ contact_zip/postal_code: chr \"98110\"\n",
      "  .. .. .. .. ..$ contributor            : chr \"K,,Zhang\\nT,,Ideker\"\n",
      "  .. .. .. .. ..$ geo_accession          : chr \"GSE40279\"\n",
      "  .. .. .. .. ..$ last_update_date       : chr \"Jul 06 2022\"\n",
      "  .. .. .. .. ..$ overall_design         : chr \"Bisulphite converted DNA from the 656 samples were hybridised to the Illumina Infinium 450k Human Methylation Beadchip\"\n",
      "  .. .. .. .. ..$ platform_id            : chr \"GPL13534\"\n",
      "  .. .. .. .. ..$ platform_taxid         : chr \"9606\"\n",
      "  .. .. .. .. ..$ pubmed_id              : chr \"23177740\"\n",
      "  .. .. .. .. ..$ relation               : chr \"Reanalyzed by: GSE207605\"\n",
      "  .. .. .. .. ..$ sample_id              : chr \"GSM989827 GSM989828 GSM989829 GSM989830 GSM989831 GSM989832 GSM989833 GSM989834 GSM989835 GSM989836 GSM989837 G\"| __truncated__\n",
      "  .. .. .. .. ..$ sample_taxid           : chr \"9606\"\n",
      "  .. .. .. .. ..$ status                 : chr \"Public on Nov 21 2012\"\n",
      "  .. .. .. .. ..$ submission_date        : chr \"Aug 21 2012\"\n",
      "  .. .. .. .. ..$ summary                : chr \"Genome wide DNA methylation profiling of individuals across a large age range. The Illumina Infinium 450k Human\"| __truncated__\n",
      "  .. .. .. .. ..$ supplementary_file     : chr \"ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE40nnn/GSE40279/suppl/GSE40279_RAW.tar\\nftp://ftp.ncbi.nlm.nih.gov/geo/\"| __truncated__\n",
      "  .. .. .. .. ..$ title                  : chr \"Genome-wide Methylation Profiles Reveal Quantitative Views of Human Aging Rates\"\n",
      "  .. .. .. .. ..$ type                   : chr \"Methylation profiling by array\"\n",
      "  .. .. .. ..@ .__classVersion__:Formal class 'Versions' [package \"Biobase\"] with 1 slot\n",
      "  .. .. .. .. .. ..@ .Data:List of 2\n",
      "  .. .. .. .. .. .. ..$ : int [1:3] 1 0 0\n",
      "  .. .. .. .. .. .. ..$ : int [1:3] 1 1 0\n",
      "  .. .. .. .. .. ..$ names: chr [1:2] \"MIAxE\" \"MIAME\"\n",
      "  .. ..@ assayData        :<environment: 0x562a5bd89a28> \n",
      "  .. ..@ phenoData        :Formal class 'AnnotatedDataFrame' [package \"Biobase\"] with 4 slots\n",
      "  .. .. .. ..@ varMetadata      :'data.frame':\t41 obs. of  1 variable:\n",
      "  .. .. .. .. ..$ labelDescription: chr [1:41] NA NA NA NA ...\n",
      "  .. .. .. ..@ data             :'data.frame':\t656 obs. of  41 variables:\n",
      "  .. .. .. .. ..$ title                  : chr [1:656] \"age 67y 1001\" \"age 89y 1002\" \"age 66y 1003\" \"age 64y 1004\" ...\n",
      "  .. .. .. .. ..$ geo_accession          : chr [1:656] \"GSM989827\" \"GSM989828\" \"GSM989829\" \"GSM989830\" ...\n",
      "  .. .. .. .. ..$ status                 : chr [1:656] \"Public on Nov 21 2012\" \"Public on Nov 21 2012\" \"Public on Nov 21 2012\" \"Public on Nov 21 2012\" ...\n",
      "  .. .. .. .. ..$ submission_date        : chr [1:656] \"Aug 21 2012\" \"Aug 21 2012\" \"Aug 21 2012\" \"Aug 21 2012\" ...\n",
      "  .. .. .. .. ..$ last_update_date       : chr [1:656] \"Nov 21 2012\" \"Nov 21 2012\" \"Nov 21 2012\" \"Nov 21 2012\" ...\n",
      "  .. .. .. .. ..$ type                   : chr [1:656] \"genomic\" \"genomic\" \"genomic\" \"genomic\" ...\n",
      "  .. .. .. .. ..$ channel_count          : chr [1:656] \"1\" \"1\" \"1\" \"1\" ...\n",
      "  .. .. .. .. ..$ source_name_ch1        : chr [1:656] \"X1001\" \"X1002\" \"X1003\" \"X1004\" ...\n",
      "  .. .. .. .. ..$ organism_ch1           : chr [1:656] \"Homo sapiens\" \"Homo sapiens\" \"Homo sapiens\" \"Homo sapiens\" ...\n",
      "  .. .. .. .. ..$ characteristics_ch1    : chr [1:656] \"age (y): 67\" \"age (y): 89\" \"age (y): 66\" \"age (y): 64\" ...\n",
      "  .. .. .. .. ..$ characteristics_ch1.1  : chr [1:656] \"source: UCSD\" \"source: UCSD\" \"source: UCSD\" \"source: UCSD\" ...\n",
      "  .. .. .. .. ..$ characteristics_ch1.2  : chr [1:656] \"plate: 1\" \"plate: 1\" \"plate: 1\" \"plate: 1\" ...\n",
      "  .. .. .. .. ..$ characteristics_ch1.3  : chr [1:656] \"gender: F\" \"gender: F\" \"gender: F\" \"gender: F\" ...\n",
      "  .. .. .. .. ..$ characteristics_ch1.4  : chr [1:656] \"ethnicity: Caucasian - European\" \"ethnicity: Caucasian - European\" \"ethnicity: Caucasian - European\" \"ethnicity: Caucasian - European\" ...\n",
      "  .. .. .. .. ..$ characteristics_ch1.5  : chr [1:656] \"tissue: whole blood\" \"tissue: whole blood\" \"tissue: whole blood\" \"tissue: whole blood\" ...\n",
      "  .. .. .. .. ..$ molecule_ch1           : chr [1:656] \"genomic DNA\" \"genomic DNA\" \"genomic DNA\" \"genomic DNA\" ...\n",
      "  .. .. .. .. ..$ extract_protocol_ch1   : chr [1:656] \"genomic DNA was extracted and purified Qiagen FlexiGene DNA kit\" \"genomic DNA was extracted and purified Qiagen FlexiGene DNA kit\" \"genomic DNA was extracted and purified Qiagen FlexiGene DNA kit\" \"genomic DNA was extracted and purified Qiagen FlexiGene DNA kit\" ...\n",
      "  .. .. .. .. ..$ label_ch1              : chr [1:656] \"Cy5 and Cy3\" \"Cy5 and Cy3\" \"Cy5 and Cy3\" \"Cy5 and Cy3\" ...\n",
      "  .. .. .. .. ..$ label_protocol_ch1     : chr [1:656] \"Standard Illumina Protocol\" \"Standard Illumina Protocol\" \"Standard Illumina Protocol\" \"Standard Illumina Protocol\" ...\n",
      "  .. .. .. .. ..$ taxid_ch1              : chr [1:656] \"9606\" \"9606\" \"9606\" \"9606\" ...\n",
      "  .. .. .. .. ..$ hyb_protocol           : chr [1:656] \"bisulphite converted DNA was amplified, fragmented and hybridised to Illumina Infinium Human Methylation450 Bea\"| __truncated__ \"bisulphite converted DNA was amplified, fragmented and hybridised to Illumina Infinium Human Methylation450 Bea\"| __truncated__ \"bisulphite converted DNA was amplified, fragmented and hybridised to Illumina Infinium Human Methylation450 Bea\"| __truncated__ \"bisulphite converted DNA was amplified, fragmented and hybridised to Illumina Infinium Human Methylation450 Bea\"| __truncated__ ...\n",
      "  .. .. .. .. ..$ scan_protocol          : chr [1:656] \"Arrays were imaged using BeadArray Reader using standard recommended Illumina scanner setting\" \"Arrays were imaged using BeadArray Reader using standard recommended Illumina scanner setting\" \"Arrays were imaged using BeadArray Reader using standard recommended Illumina scanner setting\" \"Arrays were imaged using BeadArray Reader using standard recommended Illumina scanner setting\" ...\n",
      "  .. .. .. .. ..$ description            : chr [1:656] \"whole blood\" \"whole blood\" \"whole blood\" \"whole blood\" ...\n",
      "  .. .. .. .. ..$ data_processing        : chr [1:656] \"BeadStudio software v3.2\" \"BeadStudio software v3.2\" \"BeadStudio software v3.2\" \"BeadStudio software v3.2\" ...\n",
      "  .. .. .. .. ..$ platform_id            : chr [1:656] \"GPL13534\" \"GPL13534\" \"GPL13534\" \"GPL13534\" ...\n",
      "  .. .. .. .. ..$ contact_name           : chr [1:656] \"Justin,,Guinney\" \"Justin,,Guinney\" \"Justin,,Guinney\" \"Justin,,Guinney\" ...\n",
      "  .. .. .. .. ..$ contact_email          : chr [1:656] \"justin.guinney@sagebase.org\" \"justin.guinney@sagebase.org\" \"justin.guinney@sagebase.org\" \"justin.guinney@sagebase.org\" ...\n",
      "  .. .. .. .. ..$ contact_institute      : chr [1:656] \"Sage Bionetworks\" \"Sage Bionetworks\" \"Sage Bionetworks\" \"Sage Bionetworks\" ...\n",
      "  .. .. .. .. ..$ contact_address        : chr [1:656] \"2606 E Roy St\" \"2606 E Roy St\" \"2606 E Roy St\" \"2606 E Roy St\" ...\n",
      "  .. .. .. .. ..$ contact_city           : chr [1:656] \"Seattle\" \"Seattle\" \"Seattle\" \"Seattle\" ...\n",
      "  .. .. .. .. ..$ contact_state          : chr [1:656] \"WA\" \"WA\" \"WA\" \"WA\" ...\n",
      "  .. .. .. .. ..$ contact_zip/postal_code: chr [1:656] \"98110\" \"98110\" \"98110\" \"98110\" ...\n",
      "  .. .. .. .. ..$ contact_country        : chr [1:656] \"USA\" \"USA\" \"USA\" \"USA\" ...\n",
      "  .. .. .. .. ..$ supplementary_file     : chr [1:656] \"NONE\" \"NONE\" \"NONE\" \"NONE\" ...\n",
      "  .. .. .. .. ..$ data_row_count         : chr [1:656] \"473034\" \"473034\" \"473034\" \"473034\" ...\n",
      "  .. .. .. .. ..$ age (y):ch1            : chr [1:656] \"67\" \"89\" \"66\" \"64\" ...\n",
      "  .. .. .. .. ..$ ethnicity:ch1          : chr [1:656] \"Caucasian - European\" \"Caucasian - European\" \"Caucasian - European\" \"Caucasian - European\" ...\n",
      "  .. .. .. .. ..$ gender:ch1             : chr [1:656] \"F\" \"F\" \"F\" \"F\" ...\n",
      "  .. .. .. .. ..$ plate:ch1              : chr [1:656] \"1\" \"1\" \"1\" \"1\" ...\n",
      "  .. .. .. .. ..$ source:ch1             : chr [1:656] \"UCSD\" \"UCSD\" \"UCSD\" \"UCSD\" ...\n",
      "  .. .. .. .. ..$ tissue:ch1             : chr [1:656] \"whole blood\" \"whole blood\" \"whole blood\" \"whole blood\" ...\n",
      "  .. .. .. ..@ dimLabels        : chr [1:2] \"sampleNames\" \"sampleColumns\"\n",
      "  .. .. .. ..@ .__classVersion__:Formal class 'Versions' [package \"Biobase\"] with 1 slot\n",
      "  .. .. .. .. .. ..@ .Data:List of 1\n",
      "  .. .. .. .. .. .. ..$ : int [1:3] 1 1 0\n",
      "  .. .. .. .. .. ..$ names: chr \"AnnotatedDataFrame\"\n",
      "  .. ..@ featureData      :Formal class 'AnnotatedDataFrame' [package \"Biobase\"] with 4 slots\n",
      "  .. .. .. ..@ varMetadata      :'data.frame':\t37 obs. of  3 variables:\n",
      "  .. .. .. .. ..$ Column          : chr [1:37] \"\" \"\" \"\" \"\" ...\n",
      "  .. .. .. .. ..$ Description     : chr [1:37] \"ID =\" \"Name =\" \"AddressA_ID =\" \"AlleleA_ProbeSeq =\" ...\n",
      "  .. .. .. .. ..$ labelDescription: chr [1:37] NA NA NA NA ...\n",
      "  .. .. .. ..@ data             :'data.frame':\t324057 obs. of  37 variables:\n",
      "  .. .. .. .. ..$ ID                         : chr [1:324057] \"cg00000029\" \"cg00000108\" \"cg00000109\" \"cg00000165\" ...\n",
      "  .. .. .. .. ..$ Name                       : chr [1:324057] \"cg00000029\" \"cg00000108\" \"cg00000109\" \"cg00000165\" ...\n",
      "  .. .. .. .. ..$ AddressA_ID                : int [1:324057] 14782418 12709357 59755374 12637463 12649348 18766346 43764508 62789509 16661505 11642304 ...\n",
      "  .. .. .. .. ..$ AlleleA_ProbeSeq           : chr [1:324057] \"AACTATACTAACRAAAAAATATCCAAAAAACACTAACRTATAAAAATTTC\" \"ATACAATAAAACAAACCTAAAATAATCCTAACTCCRCTATCATCCTAACC\" \"CAATACTAACAAACACATATACCCCCCCACAAATCTTAACTTCTAAATAC\" \"CAAAATCTATTAATACAATAACTTTTAATAAAACAACTAAAACACACATC\" ...\n",
      "  .. .. .. .. ..$ AddressB_ID                : int [1:324057] NA NA NA NA NA NA NA NA NA 38691301 ...\n",
      "  .. .. .. .. ..$ AlleleB_ProbeSeq           : chr [1:324057] \"\" \"\" \"\" \"\" ...\n",
      "  .. .. .. .. ..$ Infinium_Design_Type       : chr [1:324057] \"II\" \"II\" \"II\" \"II\" ...\n",
      "  .. .. .. .. ..$ Next_Base                  : chr [1:324057] \"\" \"\" \"\" \"\" ...\n",
      "  .. .. .. .. ..$ Color_Channel              : chr [1:324057] \"\" \"\" \"\" \"\" ...\n",
      "  .. .. .. .. ..$ Forward_Sequence           : chr [1:324057] \"TTTTTTAGATAAGGATATCCAGGCGATGAGGAAGTTTTACTTCTGGGAACAGCCTGGATA[CG]AAACCTTCACACGTCAGTGTCTTTTGGACATTTTCTCGTCAGTACAGCCCTGTTGAATGT\" \"TCCATTTTGAAGGAAAAAAATGAAGGCTCTGAAAGTGTAAATCGCTTACTGAAGGGCACA[CG]GCCAGGATGACAGCGGAGCCAGGATCACCCCAGGTCTGTCTCATTGCATATGTCATGGCT\" \"GCCTTAGTCCTGAATGAGCCATTTCTCTAAGAAGTCCTGGCTTCTTTTTTAATAGAGAAT[CG]TATTTAGAAGCCAAGATCTGTGGGGGGGTACATGTGCCTGTTAGTATTGCAGTTGTGCCT\" \"CTAAGTGCAGTCAGGATCTGTTAGTACAGTGGCTTTTGATGGAACAGCTGAGGCACACAT[CG]CCCGTGGCATGGACTCCGGGGCCGAACGCTCACGACCAAGACTTTTGCCCTTTTGAAATG\" ...\n",
      "  .. .. .. .. ..$ Genome_Build               : int [1:324057] 37 37 37 37 37 37 37 37 37 37 ...\n",
      "  .. .. .. .. ..$ CHR                        : chr [1:324057] \"16\" \"3\" \"3\" \"1\" ...\n",
      "  .. .. .. .. ..$ MAPINFO                    : int [1:324057] 53468112 37459206 171916037 91194674 42263294 69341139 28890100 41167802 230560793 23034447 ...\n",
      "  .. .. .. .. ..$ SourceSeq                  : chr [1:324057] \"GCTGTACTGACGAGAAAATGTCCAAAAGACACTGACGTGTGAAGGTTTCG\" \"CGGCCAGGATGACAGCGGAGCCAGGATCACCCCAGGTCTGTCTCATTGCA\" \"AATACTAACAGGCACATGTACCCCCCCACAGATCTTGGCTTCTAAATACG\" \"AGGATCTGTTAGTACAGTGGCTTTTGATGGAACAGCTGAGGCACACATCG\" ...\n",
      "  .. .. .. .. ..$ Chromosome_36              : chr [1:324057] \"16\" \"3\" \"3\" \"1\" ...\n",
      "  .. .. .. .. ..$ Coordinate_36              : chr [1:324057] \"52025613\" \"37434210\" \"173398731\" \"90967262\" ...\n",
      "  .. .. .. .. ..$ Strand                     : chr [1:324057] \"F\" \"F\" \"F\" \"R\" ...\n",
      "  .. .. .. .. ..$ Probe_SNPs                 : chr [1:324057] \"\" \"rs9857774\" \"rs9864492\" \"\" ...\n",
      "  .. .. .. .. ..$ Probe_SNPs_10              : chr [1:324057] \"\" \"\" \"\" \"\" ...\n",
      "  .. .. .. .. ..$ Random_Loci                : logi [1:324057] NA NA NA NA NA NA ...\n",
      "  .. .. .. .. ..$ Methyl27_Loci              : logi [1:324057] NA NA NA NA NA NA ...\n",
      "  .. .. .. .. ..$ UCSC_RefGene_Name          : chr [1:324057] \"RBL2\" \"C3orf35;C3orf35\" \"FNDC3B;FNDC3B\" \"\" ...\n",
      "  .. .. .. .. ..$ UCSC_RefGene_Accession     : chr [1:324057] \"NM_005611\" \"NM_178339;NM_178342\" \"NM_001135095;NM_022763\" \"\" ...\n",
      "  .. .. .. .. ..$ UCSC_RefGene_Group         : chr [1:324057] \"TSS1500\" \"Body;3'UTR\" \"Body;Body\" \"\" ...\n",
      "  .. .. .. .. ..$ UCSC_CpG_Islands_Name      : chr [1:324057] \"chr16:53468284-53469209\" \"\" \"\" \"chr1:91190489-91192804\" ...\n",
      "  .. .. .. .. ..$ Relation_to_UCSC_CpG_Island: chr [1:324057] \"N_Shore\" \"\" \"\" \"S_Shore\" ...\n",
      "  .. .. .. .. ..$ Phantom                    : chr [1:324057] \"\" \"\" \"low-CpG:173398671-173398760\" \"\" ...\n",
      "  .. .. .. .. ..$ DMR                        : chr [1:324057] \"\" \"\" \"\" \"CDMR\" ...\n",
      "  .. .. .. .. ..$ Enhancer                   : logi [1:324057] NA NA NA TRUE NA NA ...\n",
      "  .. .. .. .. ..$ HMM_Island                 : chr [1:324057] \"\" \"\" \"\" \"1:90967262-90967361\" ...\n",
      "  .. .. .. .. ..$ Regulatory_Feature_Name    : chr [1:324057] \"16:53467838-53469685\" \"\" \"\" \"\" ...\n",
      "  .. .. .. .. ..$ Regulatory_Feature_Group   : chr [1:324057] \"Promoter_Associated\" \"\" \"\" \"\" ...\n",
      "  .. .. .. .. ..$ DHS                        : logi [1:324057] TRUE NA NA NA NA NA ...\n",
      "  .. .. .. .. ..$ RANGE_START                : int [1:324057] 53468112 37459206 171916037 91194674 42263294 69341139 28890100 41167802 230560793 23034447 ...\n",
      "  .. .. .. .. ..$ RANGE_END                  : int [1:324057] 53468235 37459329 171916160 91194797 42263417 69341262 28890223 41167925 230560916 23034570 ...\n",
      "  .. .. .. .. ..$ RANGE_GB                   : chr [1:324057] \"NC_000016.9\" \"NC_000003.11\" \"NC_000003.11\" \"NC_000001.10\" ...\n",
      "  .. .. .. .. ..$ SPOT_ID                    : chr [1:324057] \"\" \"\" \"\" \"\" ...\n",
      "  .. .. .. ..@ dimLabels        : chr [1:2] \"featureNames\" \"featureColumns\"\n",
      "  .. .. .. ..@ .__classVersion__:Formal class 'Versions' [package \"Biobase\"] with 1 slot\n",
      "  .. .. .. .. .. ..@ .Data:List of 1\n",
      "  .. .. .. .. .. .. ..$ : int [1:3] 1 1 0\n",
      "  .. .. .. .. .. ..$ names: chr \"AnnotatedDataFrame\"\n",
      "  .. ..@ annotation       : chr \"GPL13534\"\n",
      "  .. ..@ protocolData     :Formal class 'AnnotatedDataFrame' [package \"Biobase\"] with 4 slots\n",
      "  .. .. .. ..@ varMetadata      :'data.frame':\t0 obs. of  1 variable:\n",
      "  .. .. .. .. ..$ labelDescription: chr(0) \n",
      "  .. .. .. ..@ data             :'data.frame':\t656 obs. of  0 variables\n",
      "  .. .. .. ..@ dimLabels        : chr [1:2] \"sampleNames\" \"sampleColumns\"\n",
      "  .. .. .. ..@ .__classVersion__:Formal class 'Versions' [package \"Biobase\"] with 1 slot\n",
      "  .. .. .. .. .. ..@ .Data:List of 1\n",
      "  .. .. .. .. .. .. ..$ : int [1:3] 1 1 0\n",
      "  .. .. .. .. .. ..$ names: chr \"AnnotatedDataFrame\"\n",
      "  .. ..@ .__classVersion__:Formal class 'Versions' [package \"Biobase\"] with 1 slot\n",
      "  .. .. .. ..@ .Data:List of 4\n",
      "  .. .. .. .. ..$ : int [1:3] 4 1 2\n",
      "  .. .. .. .. ..$ : int [1:3] 2 54 0\n",
      "  .. .. .. .. ..$ : int [1:3] 1 3 0\n",
      "  .. .. .. .. ..$ : int [1:3] 1 0 0\n",
      "  .. .. .. ..$ names: chr [1:4] \"R\" \"Biobase\" \"eSet\" \"ExpressionSet\"\n"
     ]
    }
   ],
   "source": [
    "str(gse)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#general bad cpgs\n",
    "p <- 10000 # num of \"control probes\" for calculating control probe PCs\n",
    "nonspecific_probes <- read.table(\"https://raw.githubusercontent.com/cozygene/glint/master/parsers/assets/nonspecific_probes.txt\")[,1]\n",
    "XY_probes <- read.table(\"https://raw.githubusercontent.com/cozygene/glint/master/parsers/assets/HumanMethylationSites_X_Y.txt\")[,1]\n",
    "polymorphic_probes <- read.table(\"https://raw.githubusercontent.com/cozygene/glint/master/parsers/assets/polymorphic_cpgs.txt\")[,1]\n",
    "exclude <- union(nonspecific_probes,union(XY_probes,polymorphic_probes))\n",
    "\n",
    "\n",
    "# covariates\n",
    "cov.hannum <- cbind(as.numeric(as.factor(Biobase::pData(gse[[1]])[[\"age (y):ch1\"]])), \n",
    "                    as.numeric(as.factor(Biobase::pData(gse[[1]])[[\"gender:ch1\"]])),\n",
    "                    as.numeric(as.factor(Biobase::pData(gse[[1]])[[\"plate:ch1\"]])), \n",
    "                    as.character(as.factor(Biobase::pData(gse[[1]])[[\"ethnicity:ch1\"]])))\n",
    "\n",
    "rownames(cov.hannum) <- Biobase::pData(gse[[1]])[[\"geo_accession\"]]\n",
    "colnames(cov.hannum) <- c(\"age\", \"gender\", \"plate\",\"ethnicity\")\n",
    "\n",
    "# methylaion data\n",
    "X.hannum <- Biobase::exprs(gse[[1]])\n",
    "# remove sites with missing values \n",
    "X.hannum <- X.hannum[rowSums(is.na(X.hannum)) == 0,]\n",
    "X.hannum = X.hannum[, rownames(cov.hannum)] #make sure they are the same order"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# estimate cell-type proportions using a reference-based approach \n",
    "ref <- as.matrix(EpiDISH::centDHSbloodDMC.m[,c(\"Neutro\",\"Eosino\",\"CD4T\",\"CD8T\",\"Mono\",\"B\",\"NK\")])\n",
    "W.hannum <- EpiDISH::epidish(X.hannum, ref)$estF\n",
    "hannum <- list(X=X.hannum,\n",
    "               cov=cov.hannum,\n",
    "               W=W.hannum)\n",
    "\n",
    "# calculate PCs from low variance probes, to be treated as control probes (as in Lenhe et al. 2015); here, since we don't work with IDAT files and therefore don't have actual control probes, we use sites with the least variation in the data as control probes (as in Rahmani et al. 2019).\n",
    "site.variances <- matrixStats::rowVars(hannum$X)\n",
    "names(site.variances) <- rownames(hannum$X)\n",
    "\n",
    "low.var.sites <- names(head(sort(site.variances), p))\n",
    "low.var.pca <- prcomp(t(hannum$X[low.var.sites,]), center=TRUE, scale=TRUE, rank=20)\n",
    "hannum$ctrl_pcs <- low.var.pca$x\n",
    "\n",
    "# remove polymorphic or cross-reactive sites and non-autosomal sites; exclude low variance cpgs\n",
    "low_var_sites <- names(which(site.variances<0.001))\n",
    "keep <- setdiff(rownames(hannum$X),union(low_var_sites, exclude))\n",
    "\n",
    "hannum$X <- hannum$X[keep,]\n",
    "W_colnames <- c(\"Gran\",setdiff(colnames(hannum$W), c(\"Neutro\",\"Eosino\")))\n",
    "hannum$W <- cbind(rowSums(hannum$W[,c(\"Neutro\",\"Eosino\")]), hannum$W[,setdiff(colnames(hannum$W), c(\"Neutro\",\"Eosino\"))])\n",
    "colnames(hannum$W) <- W_colnames\n",
    "hannum$W = hannum$W[, order(colMeans(hannum$W), decreasing = T)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "List of 4\n",
      " $ X       : num [1:97792, 1:656] 0.464 0.911 0.132 0.718 0.687 ...\n",
      "  ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. ..$ : chr [1:97792] \"cg00000029\" \"cg00000109\" \"cg00000165\" \"cg00000236\" ...\n",
      "  .. ..$ : chr [1:656] \"GSM989827\" \"GSM989828\" \"GSM989829\" \"GSM989830\" ...\n",
      " $ cov     : chr [1:656, 1:4] \"47\" \"69\" \"46\" \"44\" ...\n",
      "  ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. ..$ : chr [1:656] \"GSM989827\" \"GSM989828\" \"GSM989829\" \"GSM989830\" ...\n",
      "  .. ..$ : chr [1:4] \"age\" \"gender\" \"plate\" \"ethnicity\"\n",
      " $ W       : num [1:656, 1:6] 0.771 0.855 0.793 0.828 0.727 ...\n",
      "  ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. ..$ : chr [1:656] \"GSM989827\" \"GSM989828\" \"GSM989829\" \"GSM989830\" ...\n",
      "  .. ..$ : chr [1:6] \"Gran\" \"CD8T\" \"CD4T\" \"Mono\" ...\n",
      " $ ctrl_pcs: num [1:656, 1:20] -62.9 -63.1 -64.7 -83.2 -64.4 ...\n",
      "  ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. ..$ : chr [1:656] \"GSM989827\" \"GSM989828\" \"GSM989829\" \"GSM989830\" ...\n",
      "  .. ..$ : chr [1:20] \"PC1\" \"PC2\" \"PC3\" \"PC4\" ...\n"
     ]
    }
   ],
   "source": [
    "str(hannum)\n",
    "#length(unique(hannum$cov[,\"plate\"])) returns 9 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 656 × 6 of type dbl</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>Gran</th><th scope=col>CD8T</th><th scope=col>CD4T</th><th scope=col>Mono</th><th scope=col>NK</th><th scope=col>B</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>GSM989827</th><td>0.7708958</td><td>0.01788965</td><td>0.148451813</td><td>0.04867780</td><td>0.000000000</td><td>0.0140849484</td></tr>\n",
       "\t<tr><th scope=row>GSM989828</th><td>0.8549506</td><td>0.02464364</td><td>0.037986268</td><td>0.08158997</td><td>0.000000000</td><td>0.0008295391</td></tr>\n",
       "\t<tr><th scope=row>GSM989829</th><td>0.7927175</td><td>0.02192523</td><td>0.114461255</td><td>0.03476014</td><td>0.007251677</td><td>0.0288841654</td></tr>\n",
       "\t<tr><th scope=row>GSM989830</th><td>0.8275471</td><td>0.03376357</td><td>0.035015440</td><td>0.06949500</td><td>0.027238847</td><td>0.0069400686</td></tr>\n",
       "\t<tr><th scope=row>GSM989831</th><td>0.7274271</td><td>0.01106199</td><td>0.135535701</td><td>0.06405139</td><td>0.061923854</td><td>0.0000000000</td></tr>\n",
       "\t<tr><th scope=row>GSM989832</th><td>0.4420775</td><td>0.11387796</td><td>0.182838965</td><td>0.09957381</td><td>0.116889712</td><td>0.0447420686</td></tr>\n",
       "\t<tr><th scope=row>GSM989833</th><td>0.7180810</td><td>0.03231192</td><td>0.074392425</td><td>0.08256661</td><td>0.084154294</td><td>0.0084937193</td></tr>\n",
       "\t<tr><th scope=row>GSM989834</th><td>0.7728797</td><td>0.01868127</td><td>0.087740169</td><td>0.08828652</td><td>0.023351773</td><td>0.0090605932</td></tr>\n",
       "\t<tr><th scope=row>GSM989835</th><td>0.7425389</td><td>0.07350880</td><td>0.048273627</td><td>0.06506649</td><td>0.039556305</td><td>0.0310558995</td></tr>\n",
       "\t<tr><th scope=row>GSM989836</th><td>0.6633216</td><td>0.02100873</td><td>0.113860805</td><td>0.08390914</td><td>0.033711309</td><td>0.0841883954</td></tr>\n",
       "\t<tr><th scope=row>GSM989837</th><td>0.7682514</td><td>0.09569927</td><td>0.016722100</td><td>0.07278973</td><td>0.029994917</td><td>0.0165425971</td></tr>\n",
       "\t<tr><th scope=row>GSM989838</th><td>0.8270054</td><td>0.05076937</td><td>0.019936520</td><td>0.04335579</td><td>0.046466962</td><td>0.0124659382</td></tr>\n",
       "\t<tr><th scope=row>GSM989839</th><td>0.6186094</td><td>0.14929449</td><td>0.098245147</td><td>0.05423182</td><td>0.028051269</td><td>0.0515678877</td></tr>\n",
       "\t<tr><th scope=row>GSM989840</th><td>0.6835854</td><td>0.09681713</td><td>0.109769108</td><td>0.06446086</td><td>0.030875832</td><td>0.0144916975</td></tr>\n",
       "\t<tr><th scope=row>GSM989841</th><td>0.8888659</td><td>0.02327018</td><td>0.021506442</td><td>0.03311170</td><td>0.006396577</td><td>0.0268492424</td></tr>\n",
       "\t<tr><th scope=row>GSM989842</th><td>0.7432847</td><td>0.05476344</td><td>0.090316392</td><td>0.08544397</td><td>0.026191532</td><td>0.0000000000</td></tr>\n",
       "\t<tr><th scope=row>GSM989843</th><td>0.8110317</td><td>0.09623655</td><td>0.000000000</td><td>0.08030599</td><td>0.000000000</td><td>0.0124257325</td></tr>\n",
       "\t<tr><th scope=row>GSM989844</th><td>0.8990126</td><td>0.00000000</td><td>0.006478557</td><td>0.06149133</td><td>0.030300026</td><td>0.0027174675</td></tr>\n",
       "\t<tr><th scope=row>GSM989845</th><td>0.7795820</td><td>0.05882355</td><td>0.025230050</td><td>0.11205851</td><td>0.023168670</td><td>0.0011372584</td></tr>\n",
       "\t<tr><th scope=row>GSM989846</th><td>0.7342865</td><td>0.05743822</td><td>0.126203229</td><td>0.05956162</td><td>0.000000000</td><td>0.0225103819</td></tr>\n",
       "\t<tr><th scope=row>GSM989847</th><td>0.5500974</td><td>0.29378164</td><td>0.000000000</td><td>0.06544515</td><td>0.060350774</td><td>0.0303250072</td></tr>\n",
       "\t<tr><th scope=row>GSM989848</th><td>0.7325093</td><td>0.01751270</td><td>0.091474627</td><td>0.09553439</td><td>0.025695960</td><td>0.0372730510</td></tr>\n",
       "\t<tr><th scope=row>GSM989849</th><td>0.6114308</td><td>0.08961638</td><td>0.040034587</td><td>0.10142474</td><td>0.130955351</td><td>0.0265381847</td></tr>\n",
       "\t<tr><th scope=row>GSM989850</th><td>0.8624247</td><td>0.00000000</td><td>0.032941615</td><td>0.09989373</td><td>0.004739996</td><td>0.0000000000</td></tr>\n",
       "\t<tr><th scope=row>GSM989851</th><td>0.5542322</td><td>0.06790316</td><td>0.126484875</td><td>0.12865922</td><td>0.062390009</td><td>0.0603305025</td></tr>\n",
       "\t<tr><th scope=row>GSM989852</th><td>0.6792952</td><td>0.18661631</td><td>0.007302636</td><td>0.07798852</td><td>0.048797348</td><td>0.0000000000</td></tr>\n",
       "\t<tr><th scope=row>GSM989853</th><td>0.6805858</td><td>0.17394277</td><td>0.000000000</td><td>0.09437928</td><td>0.035526646</td><td>0.0155654959</td></tr>\n",
       "\t<tr><th scope=row>GSM989854</th><td>0.6719545</td><td>0.12100683</td><td>0.080348137</td><td>0.10611799</td><td>0.010539137</td><td>0.0100333603</td></tr>\n",
       "\t<tr><th scope=row>GSM989855</th><td>0.7609804</td><td>0.08516488</td><td>0.032073899</td><td>0.11813560</td><td>0.003645207</td><td>0.0000000000</td></tr>\n",
       "\t<tr><th scope=row>GSM989856</th><td>0.6475813</td><td>0.08205297</td><td>0.095602393</td><td>0.06779393</td><td>0.021498050</td><td>0.0854713823</td></tr>\n",
       "\t<tr><th scope=row>⋮</th><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><th scope=row>GSM990598</th><td>0.6660568</td><td>0.13452457</td><td>0.040591849</td><td>0.07114192</td><td>0.084353305</td><td>0.003331593</td></tr>\n",
       "\t<tr><th scope=row>GSM990599</th><td>0.6512056</td><td>0.15079371</td><td>0.041849551</td><td>0.04225625</td><td>0.063076616</td><td>0.050818297</td></tr>\n",
       "\t<tr><th scope=row>GSM990600</th><td>0.5539003</td><td>0.00000000</td><td>0.175056987</td><td>0.06793864</td><td>0.166891705</td><td>0.036212368</td></tr>\n",
       "\t<tr><th scope=row>GSM990601</th><td>0.7336796</td><td>0.07852628</td><td>0.051841739</td><td>0.06286417</td><td>0.018303277</td><td>0.054784941</td></tr>\n",
       "\t<tr><th scope=row>GSM990602</th><td>0.7085733</td><td>0.04884108</td><td>0.114788772</td><td>0.05787337</td><td>0.024898099</td><td>0.045025347</td></tr>\n",
       "\t<tr><th scope=row>GSM990603</th><td>0.7151475</td><td>0.09294272</td><td>0.089291128</td><td>0.04211546</td><td>0.040933898</td><td>0.019569276</td></tr>\n",
       "\t<tr><th scope=row>GSM990604</th><td>0.7516969</td><td>0.02266357</td><td>0.097921941</td><td>0.10363036</td><td>0.024087239</td><td>0.000000000</td></tr>\n",
       "\t<tr><th scope=row>GSM990605</th><td>0.6638234</td><td>0.09785907</td><td>0.006167925</td><td>0.10519793</td><td>0.082788919</td><td>0.044162794</td></tr>\n",
       "\t<tr><th scope=row>GSM990606</th><td>0.7243574</td><td>0.00000000</td><td>0.087023457</td><td>0.09477143</td><td>0.062487104</td><td>0.031360635</td></tr>\n",
       "\t<tr><th scope=row>GSM990607</th><td>0.8241562</td><td>0.01112584</td><td>0.078433730</td><td>0.06723407</td><td>0.019050203</td><td>0.000000000</td></tr>\n",
       "\t<tr><th scope=row>GSM990608</th><td>0.7144145</td><td>0.05093808</td><td>0.050548859</td><td>0.11207533</td><td>0.067953171</td><td>0.004070097</td></tr>\n",
       "\t<tr><th scope=row>GSM990609</th><td>0.6465180</td><td>0.12194452</td><td>0.091214745</td><td>0.09621570</td><td>0.009209077</td><td>0.034897992</td></tr>\n",
       "\t<tr><th scope=row>GSM990610</th><td>0.6013634</td><td>0.18305536</td><td>0.075046914</td><td>0.07851255</td><td>0.010100218</td><td>0.051921552</td></tr>\n",
       "\t<tr><th scope=row>GSM990611</th><td>0.5291850</td><td>0.19373300</td><td>0.145166423</td><td>0.07908819</td><td>0.040779338</td><td>0.012048005</td></tr>\n",
       "\t<tr><th scope=row>GSM990612</th><td>0.6991232</td><td>0.07861105</td><td>0.031436213</td><td>0.16029343</td><td>0.030536101</td><td>0.000000000</td></tr>\n",
       "\t<tr><th scope=row>GSM990613</th><td>0.7254532</td><td>0.13255011</td><td>0.066761018</td><td>0.06138191</td><td>0.000000000</td><td>0.013853796</td></tr>\n",
       "\t<tr><th scope=row>GSM990614</th><td>0.7976117</td><td>0.01176851</td><td>0.117136793</td><td>0.05320480</td><td>0.006368465</td><td>0.013909779</td></tr>\n",
       "\t<tr><th scope=row>GSM990615</th><td>0.7056786</td><td>0.00000000</td><td>0.164520491</td><td>0.07597615</td><td>0.051883327</td><td>0.001941419</td></tr>\n",
       "\t<tr><th scope=row>GSM990616</th><td>0.5645421</td><td>0.05555090</td><td>0.147893592</td><td>0.13256155</td><td>0.066803089</td><td>0.032648736</td></tr>\n",
       "\t<tr><th scope=row>GSM990617</th><td>0.6069207</td><td>0.13145205</td><td>0.076639885</td><td>0.08974337</td><td>0.056453236</td><td>0.038790763</td></tr>\n",
       "\t<tr><th scope=row>GSM990618</th><td>0.5569920</td><td>0.05519286</td><td>0.217482198</td><td>0.06747556</td><td>0.054608372</td><td>0.048249022</td></tr>\n",
       "\t<tr><th scope=row>GSM990619</th><td>0.5666933</td><td>0.05440621</td><td>0.143618982</td><td>0.17072429</td><td>0.025052888</td><td>0.039504329</td></tr>\n",
       "\t<tr><th scope=row>GSM990620</th><td>0.8712257</td><td>0.02420634</td><td>0.043158567</td><td>0.04808608</td><td>0.004410762</td><td>0.008912505</td></tr>\n",
       "\t<tr><th scope=row>GSM990621</th><td>0.6791967</td><td>0.00000000</td><td>0.162456362</td><td>0.05770595</td><td>0.071813755</td><td>0.028827206</td></tr>\n",
       "\t<tr><th scope=row>GSM990622</th><td>0.6299134</td><td>0.03552174</td><td>0.178403578</td><td>0.07057436</td><td>0.053725441</td><td>0.031861428</td></tr>\n",
       "\t<tr><th scope=row>GSM990623</th><td>0.7744285</td><td>0.01664749</td><td>0.111546289</td><td>0.03827185</td><td>0.026388767</td><td>0.032717054</td></tr>\n",
       "\t<tr><th scope=row>GSM990624</th><td>0.6619085</td><td>0.01711235</td><td>0.150041921</td><td>0.09027766</td><td>0.060379848</td><td>0.020279775</td></tr>\n",
       "\t<tr><th scope=row>GSM990625</th><td>0.7208502</td><td>0.09962126</td><td>0.039956925</td><td>0.11559765</td><td>0.016260644</td><td>0.007713290</td></tr>\n",
       "\t<tr><th scope=row>GSM990626</th><td>0.6797534</td><td>0.08378410</td><td>0.162458890</td><td>0.04702628</td><td>0.005902631</td><td>0.021074655</td></tr>\n",
       "\t<tr><th scope=row>GSM990627</th><td>0.5142665</td><td>0.12196326</td><td>0.073701067</td><td>0.11036511</td><td>0.110924697</td><td>0.068779344</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 656 × 6 of type dbl\n",
       "\\begin{tabular}{r|llllll}\n",
       "  & Gran & CD8T & CD4T & Mono & NK & B\\\\\n",
       "\\hline\n",
       "\tGSM989827 & 0.7708958 & 0.01788965 & 0.148451813 & 0.04867780 & 0.000000000 & 0.0140849484\\\\\n",
       "\tGSM989828 & 0.8549506 & 0.02464364 & 0.037986268 & 0.08158997 & 0.000000000 & 0.0008295391\\\\\n",
       "\tGSM989829 & 0.7927175 & 0.02192523 & 0.114461255 & 0.03476014 & 0.007251677 & 0.0288841654\\\\\n",
       "\tGSM989830 & 0.8275471 & 0.03376357 & 0.035015440 & 0.06949500 & 0.027238847 & 0.0069400686\\\\\n",
       "\tGSM989831 & 0.7274271 & 0.01106199 & 0.135535701 & 0.06405139 & 0.061923854 & 0.0000000000\\\\\n",
       "\tGSM989832 & 0.4420775 & 0.11387796 & 0.182838965 & 0.09957381 & 0.116889712 & 0.0447420686\\\\\n",
       "\tGSM989833 & 0.7180810 & 0.03231192 & 0.074392425 & 0.08256661 & 0.084154294 & 0.0084937193\\\\\n",
       "\tGSM989834 & 0.7728797 & 0.01868127 & 0.087740169 & 0.08828652 & 0.023351773 & 0.0090605932\\\\\n",
       "\tGSM989835 & 0.7425389 & 0.07350880 & 0.048273627 & 0.06506649 & 0.039556305 & 0.0310558995\\\\\n",
       "\tGSM989836 & 0.6633216 & 0.02100873 & 0.113860805 & 0.08390914 & 0.033711309 & 0.0841883954\\\\\n",
       "\tGSM989837 & 0.7682514 & 0.09569927 & 0.016722100 & 0.07278973 & 0.029994917 & 0.0165425971\\\\\n",
       "\tGSM989838 & 0.8270054 & 0.05076937 & 0.019936520 & 0.04335579 & 0.046466962 & 0.0124659382\\\\\n",
       "\tGSM989839 & 0.6186094 & 0.14929449 & 0.098245147 & 0.05423182 & 0.028051269 & 0.0515678877\\\\\n",
       "\tGSM989840 & 0.6835854 & 0.09681713 & 0.109769108 & 0.06446086 & 0.030875832 & 0.0144916975\\\\\n",
       "\tGSM989841 & 0.8888659 & 0.02327018 & 0.021506442 & 0.03311170 & 0.006396577 & 0.0268492424\\\\\n",
       "\tGSM989842 & 0.7432847 & 0.05476344 & 0.090316392 & 0.08544397 & 0.026191532 & 0.0000000000\\\\\n",
       "\tGSM989843 & 0.8110317 & 0.09623655 & 0.000000000 & 0.08030599 & 0.000000000 & 0.0124257325\\\\\n",
       "\tGSM989844 & 0.8990126 & 0.00000000 & 0.006478557 & 0.06149133 & 0.030300026 & 0.0027174675\\\\\n",
       "\tGSM989845 & 0.7795820 & 0.05882355 & 0.025230050 & 0.11205851 & 0.023168670 & 0.0011372584\\\\\n",
       "\tGSM989846 & 0.7342865 & 0.05743822 & 0.126203229 & 0.05956162 & 0.000000000 & 0.0225103819\\\\\n",
       "\tGSM989847 & 0.5500974 & 0.29378164 & 0.000000000 & 0.06544515 & 0.060350774 & 0.0303250072\\\\\n",
       "\tGSM989848 & 0.7325093 & 0.01751270 & 0.091474627 & 0.09553439 & 0.025695960 & 0.0372730510\\\\\n",
       "\tGSM989849 & 0.6114308 & 0.08961638 & 0.040034587 & 0.10142474 & 0.130955351 & 0.0265381847\\\\\n",
       "\tGSM989850 & 0.8624247 & 0.00000000 & 0.032941615 & 0.09989373 & 0.004739996 & 0.0000000000\\\\\n",
       "\tGSM989851 & 0.5542322 & 0.06790316 & 0.126484875 & 0.12865922 & 0.062390009 & 0.0603305025\\\\\n",
       "\tGSM989852 & 0.6792952 & 0.18661631 & 0.007302636 & 0.07798852 & 0.048797348 & 0.0000000000\\\\\n",
       "\tGSM989853 & 0.6805858 & 0.17394277 & 0.000000000 & 0.09437928 & 0.035526646 & 0.0155654959\\\\\n",
       "\tGSM989854 & 0.6719545 & 0.12100683 & 0.080348137 & 0.10611799 & 0.010539137 & 0.0100333603\\\\\n",
       "\tGSM989855 & 0.7609804 & 0.08516488 & 0.032073899 & 0.11813560 & 0.003645207 & 0.0000000000\\\\\n",
       "\tGSM989856 & 0.6475813 & 0.08205297 & 0.095602393 & 0.06779393 & 0.021498050 & 0.0854713823\\\\\n",
       "\t⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\tGSM990598 & 0.6660568 & 0.13452457 & 0.040591849 & 0.07114192 & 0.084353305 & 0.003331593\\\\\n",
       "\tGSM990599 & 0.6512056 & 0.15079371 & 0.041849551 & 0.04225625 & 0.063076616 & 0.050818297\\\\\n",
       "\tGSM990600 & 0.5539003 & 0.00000000 & 0.175056987 & 0.06793864 & 0.166891705 & 0.036212368\\\\\n",
       "\tGSM990601 & 0.7336796 & 0.07852628 & 0.051841739 & 0.06286417 & 0.018303277 & 0.054784941\\\\\n",
       "\tGSM990602 & 0.7085733 & 0.04884108 & 0.114788772 & 0.05787337 & 0.024898099 & 0.045025347\\\\\n",
       "\tGSM990603 & 0.7151475 & 0.09294272 & 0.089291128 & 0.04211546 & 0.040933898 & 0.019569276\\\\\n",
       "\tGSM990604 & 0.7516969 & 0.02266357 & 0.097921941 & 0.10363036 & 0.024087239 & 0.000000000\\\\\n",
       "\tGSM990605 & 0.6638234 & 0.09785907 & 0.006167925 & 0.10519793 & 0.082788919 & 0.044162794\\\\\n",
       "\tGSM990606 & 0.7243574 & 0.00000000 & 0.087023457 & 0.09477143 & 0.062487104 & 0.031360635\\\\\n",
       "\tGSM990607 & 0.8241562 & 0.01112584 & 0.078433730 & 0.06723407 & 0.019050203 & 0.000000000\\\\\n",
       "\tGSM990608 & 0.7144145 & 0.05093808 & 0.050548859 & 0.11207533 & 0.067953171 & 0.004070097\\\\\n",
       "\tGSM990609 & 0.6465180 & 0.12194452 & 0.091214745 & 0.09621570 & 0.009209077 & 0.034897992\\\\\n",
       "\tGSM990610 & 0.6013634 & 0.18305536 & 0.075046914 & 0.07851255 & 0.010100218 & 0.051921552\\\\\n",
       "\tGSM990611 & 0.5291850 & 0.19373300 & 0.145166423 & 0.07908819 & 0.040779338 & 0.012048005\\\\\n",
       "\tGSM990612 & 0.6991232 & 0.07861105 & 0.031436213 & 0.16029343 & 0.030536101 & 0.000000000\\\\\n",
       "\tGSM990613 & 0.7254532 & 0.13255011 & 0.066761018 & 0.06138191 & 0.000000000 & 0.013853796\\\\\n",
       "\tGSM990614 & 0.7976117 & 0.01176851 & 0.117136793 & 0.05320480 & 0.006368465 & 0.013909779\\\\\n",
       "\tGSM990615 & 0.7056786 & 0.00000000 & 0.164520491 & 0.07597615 & 0.051883327 & 0.001941419\\\\\n",
       "\tGSM990616 & 0.5645421 & 0.05555090 & 0.147893592 & 0.13256155 & 0.066803089 & 0.032648736\\\\\n",
       "\tGSM990617 & 0.6069207 & 0.13145205 & 0.076639885 & 0.08974337 & 0.056453236 & 0.038790763\\\\\n",
       "\tGSM990618 & 0.5569920 & 0.05519286 & 0.217482198 & 0.06747556 & 0.054608372 & 0.048249022\\\\\n",
       "\tGSM990619 & 0.5666933 & 0.05440621 & 0.143618982 & 0.17072429 & 0.025052888 & 0.039504329\\\\\n",
       "\tGSM990620 & 0.8712257 & 0.02420634 & 0.043158567 & 0.04808608 & 0.004410762 & 0.008912505\\\\\n",
       "\tGSM990621 & 0.6791967 & 0.00000000 & 0.162456362 & 0.05770595 & 0.071813755 & 0.028827206\\\\\n",
       "\tGSM990622 & 0.6299134 & 0.03552174 & 0.178403578 & 0.07057436 & 0.053725441 & 0.031861428\\\\\n",
       "\tGSM990623 & 0.7744285 & 0.01664749 & 0.111546289 & 0.03827185 & 0.026388767 & 0.032717054\\\\\n",
       "\tGSM990624 & 0.6619085 & 0.01711235 & 0.150041921 & 0.09027766 & 0.060379848 & 0.020279775\\\\\n",
       "\tGSM990625 & 0.7208502 & 0.09962126 & 0.039956925 & 0.11559765 & 0.016260644 & 0.007713290\\\\\n",
       "\tGSM990626 & 0.6797534 & 0.08378410 & 0.162458890 & 0.04702628 & 0.005902631 & 0.021074655\\\\\n",
       "\tGSM990627 & 0.5142665 & 0.12196326 & 0.073701067 & 0.11036511 & 0.110924697 & 0.068779344\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 656 × 6 of type dbl\n",
       "\n",
       "| <!--/--> | Gran | CD8T | CD4T | Mono | NK | B |\n",
       "|---|---|---|---|---|---|---|\n",
       "| GSM989827 | 0.7708958 | 0.01788965 | 0.148451813 | 0.04867780 | 0.000000000 | 0.0140849484 |\n",
       "| GSM989828 | 0.8549506 | 0.02464364 | 0.037986268 | 0.08158997 | 0.000000000 | 0.0008295391 |\n",
       "| GSM989829 | 0.7927175 | 0.02192523 | 0.114461255 | 0.03476014 | 0.007251677 | 0.0288841654 |\n",
       "| GSM989830 | 0.8275471 | 0.03376357 | 0.035015440 | 0.06949500 | 0.027238847 | 0.0069400686 |\n",
       "| GSM989831 | 0.7274271 | 0.01106199 | 0.135535701 | 0.06405139 | 0.061923854 | 0.0000000000 |\n",
       "| GSM989832 | 0.4420775 | 0.11387796 | 0.182838965 | 0.09957381 | 0.116889712 | 0.0447420686 |\n",
       "| GSM989833 | 0.7180810 | 0.03231192 | 0.074392425 | 0.08256661 | 0.084154294 | 0.0084937193 |\n",
       "| GSM989834 | 0.7728797 | 0.01868127 | 0.087740169 | 0.08828652 | 0.023351773 | 0.0090605932 |\n",
       "| GSM989835 | 0.7425389 | 0.07350880 | 0.048273627 | 0.06506649 | 0.039556305 | 0.0310558995 |\n",
       "| GSM989836 | 0.6633216 | 0.02100873 | 0.113860805 | 0.08390914 | 0.033711309 | 0.0841883954 |\n",
       "| GSM989837 | 0.7682514 | 0.09569927 | 0.016722100 | 0.07278973 | 0.029994917 | 0.0165425971 |\n",
       "| GSM989838 | 0.8270054 | 0.05076937 | 0.019936520 | 0.04335579 | 0.046466962 | 0.0124659382 |\n",
       "| GSM989839 | 0.6186094 | 0.14929449 | 0.098245147 | 0.05423182 | 0.028051269 | 0.0515678877 |\n",
       "| GSM989840 | 0.6835854 | 0.09681713 | 0.109769108 | 0.06446086 | 0.030875832 | 0.0144916975 |\n",
       "| GSM989841 | 0.8888659 | 0.02327018 | 0.021506442 | 0.03311170 | 0.006396577 | 0.0268492424 |\n",
       "| GSM989842 | 0.7432847 | 0.05476344 | 0.090316392 | 0.08544397 | 0.026191532 | 0.0000000000 |\n",
       "| GSM989843 | 0.8110317 | 0.09623655 | 0.000000000 | 0.08030599 | 0.000000000 | 0.0124257325 |\n",
       "| GSM989844 | 0.8990126 | 0.00000000 | 0.006478557 | 0.06149133 | 0.030300026 | 0.0027174675 |\n",
       "| GSM989845 | 0.7795820 | 0.05882355 | 0.025230050 | 0.11205851 | 0.023168670 | 0.0011372584 |\n",
       "| GSM989846 | 0.7342865 | 0.05743822 | 0.126203229 | 0.05956162 | 0.000000000 | 0.0225103819 |\n",
       "| GSM989847 | 0.5500974 | 0.29378164 | 0.000000000 | 0.06544515 | 0.060350774 | 0.0303250072 |\n",
       "| GSM989848 | 0.7325093 | 0.01751270 | 0.091474627 | 0.09553439 | 0.025695960 | 0.0372730510 |\n",
       "| GSM989849 | 0.6114308 | 0.08961638 | 0.040034587 | 0.10142474 | 0.130955351 | 0.0265381847 |\n",
       "| GSM989850 | 0.8624247 | 0.00000000 | 0.032941615 | 0.09989373 | 0.004739996 | 0.0000000000 |\n",
       "| GSM989851 | 0.5542322 | 0.06790316 | 0.126484875 | 0.12865922 | 0.062390009 | 0.0603305025 |\n",
       "| GSM989852 | 0.6792952 | 0.18661631 | 0.007302636 | 0.07798852 | 0.048797348 | 0.0000000000 |\n",
       "| GSM989853 | 0.6805858 | 0.17394277 | 0.000000000 | 0.09437928 | 0.035526646 | 0.0155654959 |\n",
       "| GSM989854 | 0.6719545 | 0.12100683 | 0.080348137 | 0.10611799 | 0.010539137 | 0.0100333603 |\n",
       "| GSM989855 | 0.7609804 | 0.08516488 | 0.032073899 | 0.11813560 | 0.003645207 | 0.0000000000 |\n",
       "| GSM989856 | 0.6475813 | 0.08205297 | 0.095602393 | 0.06779393 | 0.021498050 | 0.0854713823 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| GSM990598 | 0.6660568 | 0.13452457 | 0.040591849 | 0.07114192 | 0.084353305 | 0.003331593 |\n",
       "| GSM990599 | 0.6512056 | 0.15079371 | 0.041849551 | 0.04225625 | 0.063076616 | 0.050818297 |\n",
       "| GSM990600 | 0.5539003 | 0.00000000 | 0.175056987 | 0.06793864 | 0.166891705 | 0.036212368 |\n",
       "| GSM990601 | 0.7336796 | 0.07852628 | 0.051841739 | 0.06286417 | 0.018303277 | 0.054784941 |\n",
       "| GSM990602 | 0.7085733 | 0.04884108 | 0.114788772 | 0.05787337 | 0.024898099 | 0.045025347 |\n",
       "| GSM990603 | 0.7151475 | 0.09294272 | 0.089291128 | 0.04211546 | 0.040933898 | 0.019569276 |\n",
       "| GSM990604 | 0.7516969 | 0.02266357 | 0.097921941 | 0.10363036 | 0.024087239 | 0.000000000 |\n",
       "| GSM990605 | 0.6638234 | 0.09785907 | 0.006167925 | 0.10519793 | 0.082788919 | 0.044162794 |\n",
       "| GSM990606 | 0.7243574 | 0.00000000 | 0.087023457 | 0.09477143 | 0.062487104 | 0.031360635 |\n",
       "| GSM990607 | 0.8241562 | 0.01112584 | 0.078433730 | 0.06723407 | 0.019050203 | 0.000000000 |\n",
       "| GSM990608 | 0.7144145 | 0.05093808 | 0.050548859 | 0.11207533 | 0.067953171 | 0.004070097 |\n",
       "| GSM990609 | 0.6465180 | 0.12194452 | 0.091214745 | 0.09621570 | 0.009209077 | 0.034897992 |\n",
       "| GSM990610 | 0.6013634 | 0.18305536 | 0.075046914 | 0.07851255 | 0.010100218 | 0.051921552 |\n",
       "| GSM990611 | 0.5291850 | 0.19373300 | 0.145166423 | 0.07908819 | 0.040779338 | 0.012048005 |\n",
       "| GSM990612 | 0.6991232 | 0.07861105 | 0.031436213 | 0.16029343 | 0.030536101 | 0.000000000 |\n",
       "| GSM990613 | 0.7254532 | 0.13255011 | 0.066761018 | 0.06138191 | 0.000000000 | 0.013853796 |\n",
       "| GSM990614 | 0.7976117 | 0.01176851 | 0.117136793 | 0.05320480 | 0.006368465 | 0.013909779 |\n",
       "| GSM990615 | 0.7056786 | 0.00000000 | 0.164520491 | 0.07597615 | 0.051883327 | 0.001941419 |\n",
       "| GSM990616 | 0.5645421 | 0.05555090 | 0.147893592 | 0.13256155 | 0.066803089 | 0.032648736 |\n",
       "| GSM990617 | 0.6069207 | 0.13145205 | 0.076639885 | 0.08974337 | 0.056453236 | 0.038790763 |\n",
       "| GSM990618 | 0.5569920 | 0.05519286 | 0.217482198 | 0.06747556 | 0.054608372 | 0.048249022 |\n",
       "| GSM990619 | 0.5666933 | 0.05440621 | 0.143618982 | 0.17072429 | 0.025052888 | 0.039504329 |\n",
       "| GSM990620 | 0.8712257 | 0.02420634 | 0.043158567 | 0.04808608 | 0.004410762 | 0.008912505 |\n",
       "| GSM990621 | 0.6791967 | 0.00000000 | 0.162456362 | 0.05770595 | 0.071813755 | 0.028827206 |\n",
       "| GSM990622 | 0.6299134 | 0.03552174 | 0.178403578 | 0.07057436 | 0.053725441 | 0.031861428 |\n",
       "| GSM990623 | 0.7744285 | 0.01664749 | 0.111546289 | 0.03827185 | 0.026388767 | 0.032717054 |\n",
       "| GSM990624 | 0.6619085 | 0.01711235 | 0.150041921 | 0.09027766 | 0.060379848 | 0.020279775 |\n",
       "| GSM990625 | 0.7208502 | 0.09962126 | 0.039956925 | 0.11559765 | 0.016260644 | 0.007713290 |\n",
       "| GSM990626 | 0.6797534 | 0.08378410 | 0.162458890 | 0.04702628 | 0.005902631 | 0.021074655 |\n",
       "| GSM990627 | 0.5142665 | 0.12196326 | 0.073701067 | 0.11036511 | 0.110924697 | 0.068779344 |\n",
       "\n"
      ],
      "text/plain": [
       "          Gran      CD8T       CD4T        Mono       NK          B           \n",
       "GSM989827 0.7708958 0.01788965 0.148451813 0.04867780 0.000000000 0.0140849484\n",
       "GSM989828 0.8549506 0.02464364 0.037986268 0.08158997 0.000000000 0.0008295391\n",
       "GSM989829 0.7927175 0.02192523 0.114461255 0.03476014 0.007251677 0.0288841654\n",
       "GSM989830 0.8275471 0.03376357 0.035015440 0.06949500 0.027238847 0.0069400686\n",
       "GSM989831 0.7274271 0.01106199 0.135535701 0.06405139 0.061923854 0.0000000000\n",
       "GSM989832 0.4420775 0.11387796 0.182838965 0.09957381 0.116889712 0.0447420686\n",
       "GSM989833 0.7180810 0.03231192 0.074392425 0.08256661 0.084154294 0.0084937193\n",
       "GSM989834 0.7728797 0.01868127 0.087740169 0.08828652 0.023351773 0.0090605932\n",
       "GSM989835 0.7425389 0.07350880 0.048273627 0.06506649 0.039556305 0.0310558995\n",
       "GSM989836 0.6633216 0.02100873 0.113860805 0.08390914 0.033711309 0.0841883954\n",
       "GSM989837 0.7682514 0.09569927 0.016722100 0.07278973 0.029994917 0.0165425971\n",
       "GSM989838 0.8270054 0.05076937 0.019936520 0.04335579 0.046466962 0.0124659382\n",
       "GSM989839 0.6186094 0.14929449 0.098245147 0.05423182 0.028051269 0.0515678877\n",
       "GSM989840 0.6835854 0.09681713 0.109769108 0.06446086 0.030875832 0.0144916975\n",
       "GSM989841 0.8888659 0.02327018 0.021506442 0.03311170 0.006396577 0.0268492424\n",
       "GSM989842 0.7432847 0.05476344 0.090316392 0.08544397 0.026191532 0.0000000000\n",
       "GSM989843 0.8110317 0.09623655 0.000000000 0.08030599 0.000000000 0.0124257325\n",
       "GSM989844 0.8990126 0.00000000 0.006478557 0.06149133 0.030300026 0.0027174675\n",
       "GSM989845 0.7795820 0.05882355 0.025230050 0.11205851 0.023168670 0.0011372584\n",
       "GSM989846 0.7342865 0.05743822 0.126203229 0.05956162 0.000000000 0.0225103819\n",
       "GSM989847 0.5500974 0.29378164 0.000000000 0.06544515 0.060350774 0.0303250072\n",
       "GSM989848 0.7325093 0.01751270 0.091474627 0.09553439 0.025695960 0.0372730510\n",
       "GSM989849 0.6114308 0.08961638 0.040034587 0.10142474 0.130955351 0.0265381847\n",
       "GSM989850 0.8624247 0.00000000 0.032941615 0.09989373 0.004739996 0.0000000000\n",
       "GSM989851 0.5542322 0.06790316 0.126484875 0.12865922 0.062390009 0.0603305025\n",
       "GSM989852 0.6792952 0.18661631 0.007302636 0.07798852 0.048797348 0.0000000000\n",
       "GSM989853 0.6805858 0.17394277 0.000000000 0.09437928 0.035526646 0.0155654959\n",
       "GSM989854 0.6719545 0.12100683 0.080348137 0.10611799 0.010539137 0.0100333603\n",
       "GSM989855 0.7609804 0.08516488 0.032073899 0.11813560 0.003645207 0.0000000000\n",
       "GSM989856 0.6475813 0.08205297 0.095602393 0.06779393 0.021498050 0.0854713823\n",
       "⋮         ⋮         ⋮          ⋮           ⋮          ⋮           ⋮           \n",
       "GSM990598 0.6660568 0.13452457 0.040591849 0.07114192 0.084353305 0.003331593 \n",
       "GSM990599 0.6512056 0.15079371 0.041849551 0.04225625 0.063076616 0.050818297 \n",
       "GSM990600 0.5539003 0.00000000 0.175056987 0.06793864 0.166891705 0.036212368 \n",
       "GSM990601 0.7336796 0.07852628 0.051841739 0.06286417 0.018303277 0.054784941 \n",
       "GSM990602 0.7085733 0.04884108 0.114788772 0.05787337 0.024898099 0.045025347 \n",
       "GSM990603 0.7151475 0.09294272 0.089291128 0.04211546 0.040933898 0.019569276 \n",
       "GSM990604 0.7516969 0.02266357 0.097921941 0.10363036 0.024087239 0.000000000 \n",
       "GSM990605 0.6638234 0.09785907 0.006167925 0.10519793 0.082788919 0.044162794 \n",
       "GSM990606 0.7243574 0.00000000 0.087023457 0.09477143 0.062487104 0.031360635 \n",
       "GSM990607 0.8241562 0.01112584 0.078433730 0.06723407 0.019050203 0.000000000 \n",
       "GSM990608 0.7144145 0.05093808 0.050548859 0.11207533 0.067953171 0.004070097 \n",
       "GSM990609 0.6465180 0.12194452 0.091214745 0.09621570 0.009209077 0.034897992 \n",
       "GSM990610 0.6013634 0.18305536 0.075046914 0.07851255 0.010100218 0.051921552 \n",
       "GSM990611 0.5291850 0.19373300 0.145166423 0.07908819 0.040779338 0.012048005 \n",
       "GSM990612 0.6991232 0.07861105 0.031436213 0.16029343 0.030536101 0.000000000 \n",
       "GSM990613 0.7254532 0.13255011 0.066761018 0.06138191 0.000000000 0.013853796 \n",
       "GSM990614 0.7976117 0.01176851 0.117136793 0.05320480 0.006368465 0.013909779 \n",
       "GSM990615 0.7056786 0.00000000 0.164520491 0.07597615 0.051883327 0.001941419 \n",
       "GSM990616 0.5645421 0.05555090 0.147893592 0.13256155 0.066803089 0.032648736 \n",
       "GSM990617 0.6069207 0.13145205 0.076639885 0.08974337 0.056453236 0.038790763 \n",
       "GSM990618 0.5569920 0.05519286 0.217482198 0.06747556 0.054608372 0.048249022 \n",
       "GSM990619 0.5666933 0.05440621 0.143618982 0.17072429 0.025052888 0.039504329 \n",
       "GSM990620 0.8712257 0.02420634 0.043158567 0.04808608 0.004410762 0.008912505 \n",
       "GSM990621 0.6791967 0.00000000 0.162456362 0.05770595 0.071813755 0.028827206 \n",
       "GSM990622 0.6299134 0.03552174 0.178403578 0.07057436 0.053725441 0.031861428 \n",
       "GSM990623 0.7744285 0.01664749 0.111546289 0.03827185 0.026388767 0.032717054 \n",
       "GSM990624 0.6619085 0.01711235 0.150041921 0.09027766 0.060379848 0.020279775 \n",
       "GSM990625 0.7208502 0.09962126 0.039956925 0.11559765 0.016260644 0.007713290 \n",
       "GSM990626 0.6797534 0.08378410 0.162458890 0.04702628 0.005902631 0.021074655 \n",
       "GSM990627 0.5142665 0.12196326 0.073701067 0.11036511 0.110924697 0.068779344 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "hannum$W"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "saveRDS(hannum, file.path(data.dir, \"hannum.rds\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Reinius"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Found 1 file(s)\n",
      "\n",
      "GSE35069_series_matrix.txt.gz\n",
      "\n",
      "Using locally cached version: ../Data/Methylation/Purified-Reinius//GSE35069_series_matrix.txt.gz\n",
      "\n",
      "Using locally cached version of GPL13534 found here:\n",
      "../Data/Methylation/Purified-Reinius//GPL13534.soft.gz \n",
      "\n"
     ]
    }
   ],
   "source": [
    "Sys.setenv(\"VROOM_CONNECTION_SIZE\" = 131072 * 10)\n",
    "gse <- GEOquery::getGEO(\"GSE35069\", destdir = data.dir, GSEMatrix = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# covars\n",
    "cov.reinius <- cbind(as.character(Biobase::pData(gse[[1]])[[\"title\"]]),\n",
    "                     as.character(Biobase::pData(gse[[1]])[[\"source_name_ch1\"]]),\n",
    "                     as.character(Biobase::pData(gse[[1]])[[\"gender:ch1\"]]))\n",
    "rownames(cov.reinius) <- Biobase::pData(gse[[1]])[[\"geo_accession\"]]\n",
    "colnames(cov.reinius) <- c(\"sample.id\", \"tissue\", \"gender\")\n",
    "cov.reinius[,\"sample.id\"] = vapply(1:nrow(cov.reinius), \n",
    "                                   function(i) paste0(\"sample.\",strsplit(cov.reinius[i,\"sample.id\"], \"_\")[[1]][2]), character(1))\n",
    "\n",
    "#same name used in EpiDISH\n",
    "cov.reinius[cov.reinius[,\"tissue\"] == \"Granulocytes\",    \"tissue\"] = \"Gran\"\n",
    "cov.reinius[cov.reinius[,\"tissue\"] == \"CD4+ T cells\",    \"tissue\"] = \"CD4T\"\n",
    "cov.reinius[cov.reinius[,\"tissue\"] == \"CD8+ T cells\",    \"tissue\"] = \"CD8T\"\n",
    "cov.reinius[cov.reinius[,\"tissue\"] == \"CD14+ Monocytes\", \"tissue\"] = \"Mono\"\n",
    "cov.reinius[cov.reinius[,\"tissue\"] == \"CD19+ B cells\",   \"tissue\"] = \"B\"\n",
    "cov.reinius[cov.reinius[,\"tissue\"] == \"CD56+ NK cells\",  \"tissue\"] = \"NK\"\n",
    "\n",
    "# methylaion data\n",
    "X.reinius <- Biobase::exprs(gse[[1]])\n",
    "X.reinius = X.reinius[, rownames(cov.reinius)] #same order as covars\n",
    "\n",
    "# remove sites with missing values \n",
    "X.reinius <- X.reinius[rowSums(is.na(X.reinius)) == 0,]\n",
    "\n",
    "source.ids  = c(\"Gran\", \"CD4T\", \"CD8T\", \"Mono\", \"B\", \"NK\")\n",
    "feature.ids = rownames(X.reinius)\n",
    "sample.ids  = sort(unique(cov.reinius[,\"sample.id\"]))\n",
    "\n",
    "k = length(source.ids)\n",
    "m = length(feature.ids)\n",
    "n = length(sample.ids)\n",
    "\n",
    "Z = array(0, c(k,m,n))\n",
    "dimnames(Z)[[1]] = source.ids\n",
    "dimnames(Z)[[2]] = feature.ids\n",
    "dimnames(Z)[[3]] = sample.ids\n",
    "\n",
    "for (source.id in source.ids){\n",
    "    for(sample.id in sample.ids){\n",
    "        gse.sample.id = rownames(cov.reinius[cov.reinius[,\"tissue\"] == source.id & cov.reinius[,\"sample.id\"] == sample.id, ,drop = F])\n",
    "        Z[source.id,,sample.id] = X.reinius[feature.ids, gse.sample.id] \n",
    "    }\n",
    "}\n",
    "\n",
    "# just the mixture X \n",
    "X = matrix(0,m,n)\n",
    "rownames(X) = feature.ids\n",
    "colnames(X) = sample.ids\n",
    "\n",
    "gse.sample.ids = c()\n",
    "for(sample.id in sample.ids){\n",
    "    gse.sample.id = rownames(cov.reinius[cov.reinius[,\"tissue\"] == \"Whole blood\" & cov.reinius[,\"sample.id\"] == sample.id, ,drop = F])\n",
    "    gse.sample.ids = c(gse.sample.ids, gse.sample.id)\n",
    "    X[feature.ids,sample.id] = X.reinius[feature.ids, gse.sample.id] \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# estimate cell-type proportions using a reference-based approach \n",
    "ref <- as.matrix(EpiDISH::centDHSbloodDMC.m[,c(\"Neutro\",\"Eosino\",\"CD4T\",\"CD8T\",\"Mono\",\"B\",\"NK\")])\n",
    "W.reinius <- EpiDISH::epidish(X, ref)$estF\n",
    "W_colnames <- c(\"Gran\",setdiff(colnames(W.reinius), c(\"Neutro\",\"Eosino\")))\n",
    "W.reinius <- cbind(rowSums(W.reinius[,c(\"Neutro\",\"Eosino\")]), W.reinius[,setdiff(colnames(W.reinius), c(\"Neutro\",\"Eosino\"))])\n",
    "colnames(W.reinius) <- W_colnames\n",
    "W.reinius = W.reinius[, order(colMeans(W.reinius), decreasing = T)]\n",
    "\n",
    "reinius <- list(X   = X,\n",
    "                cov = cov.reinius[gse.sample.ids, \"gender\", drop = F],\n",
    "                W   = W.reinius,\n",
    "                Z   = Z)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "List of 4\n",
      " $ X  : num [1:459227, 1:6] 0.541 0.964 0.794 0.22 0.829 ...\n",
      "  ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. ..$ : chr [1:459227] \"cg00000029\" \"cg00000108\" \"cg00000109\" \"cg00000165\" ...\n",
      "  .. ..$ : chr [1:6] \"sample.1\" \"sample.2\" \"sample.3\" \"sample.4\" ...\n",
      " $ cov: chr [1:6, 1] \"Male\" \"Male\" \"Male\" \"Male\" ...\n",
      "  ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. ..$ : chr [1:6] \"GSM861635\" \"GSM861636\" \"GSM861637\" \"GSM861638\" ...\n",
      "  .. ..$ : chr \"gender\"\n",
      " $ W  : num [1:6, 1:6] 0.532 0.59 0.665 0.617 0.775 ...\n",
      "  ..- attr(*, \"dimnames\")=List of 2\n",
      "  .. ..$ : chr [1:6] \"sample.1\" \"sample.2\" \"sample.3\" \"sample.4\" ...\n",
      "  .. ..$ : chr [1:6] \"Gran\" \"CD4T\" \"CD8T\" \"Mono\" ...\n",
      " $ Z  : num [1:6, 1:459227, 1:6] 0.45 0.654 0.701 0.472 0.615 ...\n",
      "  ..- attr(*, \"dimnames\")=List of 3\n",
      "  .. ..$ : chr [1:6] \"Gran\" \"CD4T\" \"CD8T\" \"Mono\" ...\n",
      "  .. ..$ : chr [1:459227] \"cg00000029\" \"cg00000108\" \"cg00000109\" \"cg00000165\" ...\n",
      "  .. ..$ : chr [1:6] \"sample.1\" \"sample.2\" \"sample.3\" \"sample.4\" ...\n"
     ]
    }
   ],
   "source": [
    "str(reinius)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A matrix: 6 × 6 of type dbl</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>Gran</th><th scope=col>CD4T</th><th scope=col>CD8T</th><th scope=col>Mono</th><th scope=col>NK</th><th scope=col>B</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>sample.1</th><td>0.5319253</td><td>0.09947401</td><td>0.14015393</td><td>0.07182887</td><td>0.12835663</td><td>0.02826126</td></tr>\n",
       "\t<tr><th scope=row>sample.2</th><td>0.5895850</td><td>0.15851072</td><td>0.01399476</td><td>0.11339855</td><td>0.07121728</td><td>0.05329369</td></tr>\n",
       "\t<tr><th scope=row>sample.3</th><td>0.6652398</td><td>0.10685107</td><td>0.11893223</td><td>0.06025880</td><td>0.02113722</td><td>0.02758086</td></tr>\n",
       "\t<tr><th scope=row>sample.4</th><td>0.6165684</td><td>0.14888277</td><td>0.07321796</td><td>0.04857812</td><td>0.03387988</td><td>0.07887290</td></tr>\n",
       "\t<tr><th scope=row>sample.5</th><td>0.7747228</td><td>0.08291011</td><td>0.04804910</td><td>0.06724337</td><td>0.00000000</td><td>0.02707467</td></tr>\n",
       "\t<tr><th scope=row>sample.6</th><td>0.6511645</td><td>0.12181533</td><td>0.09518408</td><td>0.05959220</td><td>0.03114555</td><td>0.04109829</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A matrix: 6 × 6 of type dbl\n",
       "\\begin{tabular}{r|llllll}\n",
       "  & Gran & CD4T & CD8T & Mono & NK & B\\\\\n",
       "\\hline\n",
       "\tsample.1 & 0.5319253 & 0.09947401 & 0.14015393 & 0.07182887 & 0.12835663 & 0.02826126\\\\\n",
       "\tsample.2 & 0.5895850 & 0.15851072 & 0.01399476 & 0.11339855 & 0.07121728 & 0.05329369\\\\\n",
       "\tsample.3 & 0.6652398 & 0.10685107 & 0.11893223 & 0.06025880 & 0.02113722 & 0.02758086\\\\\n",
       "\tsample.4 & 0.6165684 & 0.14888277 & 0.07321796 & 0.04857812 & 0.03387988 & 0.07887290\\\\\n",
       "\tsample.5 & 0.7747228 & 0.08291011 & 0.04804910 & 0.06724337 & 0.00000000 & 0.02707467\\\\\n",
       "\tsample.6 & 0.6511645 & 0.12181533 & 0.09518408 & 0.05959220 & 0.03114555 & 0.04109829\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A matrix: 6 × 6 of type dbl\n",
       "\n",
       "| <!--/--> | Gran | CD4T | CD8T | Mono | NK | B |\n",
       "|---|---|---|---|---|---|---|\n",
       "| sample.1 | 0.5319253 | 0.09947401 | 0.14015393 | 0.07182887 | 0.12835663 | 0.02826126 |\n",
       "| sample.2 | 0.5895850 | 0.15851072 | 0.01399476 | 0.11339855 | 0.07121728 | 0.05329369 |\n",
       "| sample.3 | 0.6652398 | 0.10685107 | 0.11893223 | 0.06025880 | 0.02113722 | 0.02758086 |\n",
       "| sample.4 | 0.6165684 | 0.14888277 | 0.07321796 | 0.04857812 | 0.03387988 | 0.07887290 |\n",
       "| sample.5 | 0.7747228 | 0.08291011 | 0.04804910 | 0.06724337 | 0.00000000 | 0.02707467 |\n",
       "| sample.6 | 0.6511645 | 0.12181533 | 0.09518408 | 0.05959220 | 0.03114555 | 0.04109829 |\n",
       "\n"
      ],
      "text/plain": [
       "         Gran      CD4T       CD8T       Mono       NK         B         \n",
       "sample.1 0.5319253 0.09947401 0.14015393 0.07182887 0.12835663 0.02826126\n",
       "sample.2 0.5895850 0.15851072 0.01399476 0.11339855 0.07121728 0.05329369\n",
       "sample.3 0.6652398 0.10685107 0.11893223 0.06025880 0.02113722 0.02758086\n",
       "sample.4 0.6165684 0.14888277 0.07321796 0.04857812 0.03387988 0.07887290\n",
       "sample.5 0.7747228 0.08291011 0.04804910 0.06724337 0.00000000 0.02707467\n",
       "sample.6 0.6511645 0.12181533 0.09518408 0.05959220 0.03114555 0.04109829"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "reinius$W"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "saveRDS(reinius, file.path(data.dir, \"reinius.rds\"))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Seurat-V4",
   "language": "R",
   "name": "ir41"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
