<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Unico-Tutorial • Unico</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Unico-Tutorial">
<meta property="og:description" content="Unico">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Unico</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Unico-Tutorial.html">Unico-Tutorial</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/cozygene/Unico/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Unico-Tutorial</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/cozygene/Unico/blob/HEAD/vignettes/Unico-Tutorial.Rmd" class="external-link"><code>vignettes/Unico-Tutorial.Rmd</code></a></small>
      <div class="hidden name"><code>Unico-Tutorial.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="load-the-unico-package-and">Load the Unico package and<a class="anchor" aria-label="anchor" href="#load-the-unico-package-and"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://cozygene.github.io/Unico/" class="external-link">Unico</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/HenrikBengtsson/matrixStats" class="external-link">matrixStats</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">#For visualization in this vignette</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ggplot2"</span>,<span class="st">"ggpubr"</span>,<span class="st">"hexbin"</span>, <span class="st">"egg"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/source.html" class="external-link">source</a></span><span class="op">(</span><span class="st">"https://github.com/cozygene/Unico/raw/main/vignettes/vignetts.utils.r"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="deconvolution-on-pseudobulk-rna-expression-data">Deconvolution on pseudobulk RNA expression data<a class="anchor" aria-label="anchor" href="#deconvolution-on-pseudobulk-rna-expression-data"></a>
</h2>
<p>Let’s first load a small simulated pseudo-bulk PBMC expression
profiles using data from the Stephenson et al, where the hidden
cell-type specific expression profile is also known. We can obtain this
toy dataset with the following command</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data_path</span> <span class="op">&lt;-</span> <span class="st">"./"</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">data_path</span>, <span class="st">"pbmc.rds"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="st">"https://github.com/cozygene/Unico/raw/main/vignettes/pbmc.rds"</span>, <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">data_path</span>,<span class="st">"pbmc.rds"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="va">sim.data</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">data_path</span>,<span class="st">"pbmc.rds"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>We use the <code>Unico</code> function to fit the Unico model. In
this simple simulation, we ignore covariates that could affect the
expression profile at cell-type level (C1) and tissue-level (C2). Once
we have all the necessary model parameters, we can ask Unico to
explicitly infer the hidden cell-types by genes by samples 3D tensor
with function <code>tensor</code>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">unico.res</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#parameter learning </span></span>
<span><span class="va">unico.res</span><span class="op">$</span><span class="va">params.hat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Unico.html">Unico</a></span><span class="op">(</span><span class="va">sim.data</span><span class="op">$</span><span class="va">X</span>, <span class="va">sim.data</span><span class="op">$</span><span class="va">W</span>, C1 <span class="op">=</span> <span class="cn">NULL</span>, C2 <span class="op">=</span> <span class="cn">NULL</span>, parallel <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#tensor </span></span>
<span><span class="va">unico.res</span><span class="op">$</span><span class="va">Z.hat</span> <span class="op">=</span> <span class="fu"><a href="../reference/tensor.html">tensor</a></span><span class="op">(</span><span class="va">sim.data</span><span class="op">$</span><span class="va">X</span>, W <span class="op">=</span> <span class="va">sim.data</span><span class="op">$</span><span class="va">W</span>, C1 <span class="op">=</span> <span class="cn">NULL</span>, C2 <span class="op">=</span> <span class="cn">NULL</span>, </span>
<span>                         <span class="va">unico.res</span><span class="op">$</span><span class="va">params.hat</span>, parallel <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>To evaluate the performance of Unico, we now check per gene, at
cell-type resolution, what is the correlation between our inferred
tensor and the ground truth cell-type-level profiles across all samples.
We repeat this process for all genes.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># evaluate tensor performance on features with variation </span></span>
<span><span class="va">unico.res</span><span class="op">$</span><span class="va">Z.corrs</span> <span class="op">=</span> <span class="fu">calc_Z_corrs</span><span class="op">(</span>Z.true <span class="op">=</span> <span class="va">sim.data</span><span class="op">$</span><span class="va">Z.scale</span>,</span>
<span>                                 Z.hat  <span class="op">=</span> <span class="va">unico.res</span><span class="op">$</span><span class="va">Z.hat</span>, </span>
<span>                                 eval.feature.source <span class="op">=</span> <span class="va">sim.data</span><span class="op">$</span><span class="va">variable.feature.source</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/matrixStats/man/rowMedians.html" class="external-link">colMedians</a></span><span class="op">(</span><span class="va">unico.res</span><span class="op">$</span><span class="va">Z.corrs</span><span class="op">)</span></span></code></pre></div>
<p>The median correlation (across all genes) for each of the cell types
are: 0.86, 0.77, 0.79, 0.58, 0.66 for CD4 T cells, NK cells, CD8 T
cells, Monocytes, B cells respectively.</p>
<p>To visualize the result, we randomly pick a gene (in the top 25% low
entropy quantile) and visualize the estimated and the ground truth
expression level at an arbitrary cell type (e.g. third most
abundant)</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">low.gene</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sim.data</span><span class="op">$</span><span class="va">params</span><span class="op">$</span><span class="va">entropies</span><span class="op">[</span><span class="va">sim.data</span><span class="op">$</span><span class="va">params</span><span class="op">$</span><span class="va">entropies</span> <span class="op">&lt;</span> <span class="fu"><a href="https://rdrr.io/r/stats/quantile.html" class="external-link">quantile</a></span><span class="op">(</span><span class="va">sim.data</span><span class="op">$</span><span class="va">params</span><span class="op">$</span><span class="va">entropies</span>, <span class="fl">0.25</span><span class="op">)</span>, ,drop<span class="op">=</span> <span class="cn">F</span><span class="op">]</span><span class="op">)</span>, <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">sim.data</span><span class="op">$</span><span class="va">Z</span><span class="op">[</span><span class="fl">3</span>,<span class="va">low.gene</span>, <span class="op">]</span>, <span class="va">unico.res</span><span class="op">$</span><span class="va">Z.hat</span><span class="op">[</span><span class="fl">3</span>,<span class="va">low.gene</span>, <span class="op">]</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/graphics/title.html" class="external-link">title</a></span> <span class="op">(</span><span class="va">low.gene</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="tensor.cor.png" alt="Correlation between estimated source-specific profile and the ground truth" width="400px"><p class="caption">
Correlation between estimated source-specific profile and the ground
truth
</p>
</div>
</div>
<div class="section level2">
<h2 id="cell-type-level-differential-methylation-analysis">Cell-type level differential methylation analysis<a class="anchor" aria-label="anchor" href="#cell-type-level-differential-methylation-analysis"></a>
</h2>
<p>Unico also offers association testing at cell-type resolution. Let’s
look at two publicly available bulk methylation datasets collected on
whole blood. Instead of looking at close to 450k CpG sites genmone-wide,
in this tutorial, we have selected a subset of just 3000 arbitrary CpG
sites spread across the autosomes. In addition, we collected covariates
reported by the original author and computed 20 components that are
expected to capture technical variation in the data. These components
were constructed by calculating the first 20 principal components from a
set of 10,000 control probes that are not expected to capture biological
variation but only technical variability in an approach similar to the
one suggested by Lehne et al. Since we did not use the raw IDAT files,
we considered 10,000 sites with the lowest variance in the data as
control probes. You can obtain a pre-processed version with the
following command.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data_path</span> <span class="op">&lt;-</span> <span class="st">"./"</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">data_path</span>, <span class="st">"liu.rds"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="st">"https://github.com/cozygene/Unico/raw/main/vignettes/liu.rds"</span>, <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">data_path</span>,<span class="st">"liu.rds"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">data_path</span>, <span class="st">"hannum.rds"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="st">"https://github.com/cozygene/Unico/raw/main/vignettes/hannum.rds"</span>, <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">data_path</span>,<span class="st">"hannum.rds"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="va">liu</span>    <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">data_path</span>,<span class="st">"liu.rds"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">hannum</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">data_path</span>,<span class="st">"hannum.rds"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>We similarly use the <code>Unico</code> function to fit the model on
Liu et al.’s data. Unlike in the gene expression simulation, we provide
additional covariates that could affect the observed methylation
profile. Specifically, we set cell-type level covariates <code>C1</code>
to age, sex, disease status and smoking status, and tissue-level
covariates C2 to technical variation components. Once we have all the
necessary model parameters, we can ask Unico to perform parametric
association testing under the assumption that methylation levels are
normally distributed.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">source.ids</span>   <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">liu</span><span class="op">$</span><span class="va">W</span><span class="op">)</span></span>
<span><span class="va">n</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">liu</span><span class="op">$</span><span class="va">X</span><span class="op">)</span></span>
<span><span class="va">m</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">liu</span><span class="op">$</span><span class="va">X</span><span class="op">)</span></span>
<span><span class="va">k</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">liu</span><span class="op">$</span><span class="va">W</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">unico.liu</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">unico.liu</span><span class="op">$</span><span class="va">params.hat</span> <span class="op">=</span> <span class="fu"><a href="../reference/Unico.html">Unico</a></span><span class="op">(</span>X  <span class="op">=</span> <span class="va">liu</span><span class="op">$</span><span class="va">X</span>, W <span class="op">=</span> <span class="va">liu</span><span class="op">$</span><span class="va">W</span>,</span>
<span>                             C1 <span class="op">=</span> <span class="va">liu</span><span class="op">$</span><span class="va">cov</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"age"</span>, <span class="st">"sex"</span>, <span class="st">"disease"</span>,<span class="st">"smoking"</span><span class="op">)</span><span class="op">]</span>,</span>
<span>                             C2 <span class="op">=</span> <span class="va">liu</span><span class="op">$</span><span class="va">ctrl_pcs</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">unico.liu</span><span class="op">$</span><span class="va">params.hat</span> <span class="op">=</span> <span class="fu"><a href="../reference/association_parametric.html">association_parametric</a></span><span class="op">(</span>X <span class="op">=</span> <span class="va">liu</span><span class="op">$</span><span class="va">X</span>, <span class="va">unico.liu</span><span class="op">$</span><span class="va">params.hat</span><span class="op">)</span></span></code></pre></div>
<p>Now, perform similar model fitting and association testing on an
independent data set (Hannum et.al.).</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">unico.hannum</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">unico.hannum</span><span class="op">$</span><span class="va">params.hat</span> <span class="op">=</span> <span class="fu"><a href="../reference/Unico.html">Unico</a></span><span class="op">(</span>X  <span class="op">=</span> <span class="va">hannum</span><span class="op">$</span><span class="va">X</span>, W <span class="op">=</span> <span class="va">hannum</span><span class="op">$</span><span class="va">W</span>,</span>
<span>                                C1 <span class="op">=</span> <span class="va">hannum</span><span class="op">$</span><span class="va">cov</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"age"</span>, <span class="st">"sex"</span>, <span class="st">"ethnicity"</span><span class="op">)</span><span class="op">]</span>,</span>
<span>                                C2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">hannum</span><span class="op">$</span><span class="va">ctrl_pcs</span>, <span class="va">hannum</span><span class="op">$</span><span class="va">cov</span><span class="op">[</span>,<span class="st">"plate"</span>, drop <span class="op">=</span> <span class="cn">F</span><span class="op">]</span><span class="op">)</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">unico.hannum</span><span class="op">$</span><span class="va">params.hat</span> <span class="op">=</span> <span class="fu"><a href="../reference/association_parametric.html">association_parametric</a></span><span class="op">(</span>X <span class="op">=</span> <span class="va">hannum</span><span class="op">$</span><span class="va">X</span>, <span class="va">unico.hannum</span><span class="op">$</span><span class="va">params.hat</span><span class="op">)</span></span></code></pre></div>
<p>Unico reports both the marginal test (T-test) on the effect size of
each covariate at cell type resolution as well as a joint test (partial
F-test) that evaluates the joint effects of each of the covariate on the
mixture profile. Let’s first look at the reported marginal association
testing results:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">liu.marg.pvals</span>    <span class="op">=</span> <span class="va">unico.liu</span><span class="op">$</span><span class="va">params.hat</span><span class="op">$</span><span class="va">parametric</span><span class="op">$</span><span class="va">gammas_hat_pvals</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">source.ids</span>, <span class="st">".age"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">hannum.marg.pvals</span> <span class="op">=</span> <span class="va">unico.hannum</span><span class="op">$</span><span class="va">params.hat</span><span class="op">$</span><span class="va">parametric</span><span class="op">$</span><span class="va">gammas_hat_pvals</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">source.ids</span>, <span class="st">".age"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">liu.marg.pvals</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">/</span><span class="op">(</span><span class="va">m</span><span class="op">*</span><span class="va">k</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">hannum.marg.pvals</span><span class="op">[</span><span class="va">liu.marg.pvals</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">/</span><span class="op">(</span><span class="va">m</span><span class="op">*</span><span class="va">k</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p>Using Liu et al.’s data as discovery, we have 4 CpG-cell-type pairs
crossing the Bonferroni adjustment (on both number of CpG sites and
number of cell types under test). All of which are replicated in our
validation data set (Hannum et al.).</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">qq_age_g</span> <span class="op">=</span> <span class="fu">plot_qq</span><span class="op">(</span>pvals_mat <span class="op">=</span> <span class="va">liu.marg.pvals</span>,</span>
<span>                   labels <span class="op">=</span> <span class="va">source.ids</span>, </span>
<span>                   ggarrange.nrow <span class="op">=</span> <span class="fl">1</span>, ggarrange.ncol <span class="op">=</span> <span class="va">k</span>, </span>
<span>                   alpha <span class="op">=</span> <span class="fl">0.5</span>, text.size <span class="op">=</span> <span class="fl">20</span>,</span>
<span>                   title <span class="op">=</span> <span class="st">"Parametric association testing (age) at cell-type resolution"</span><span class="op">)</span></span>
<span><span class="va">qq_age_g</span></span></code></pre></div>
<div class="figure">
<img src="qq_age.png" alt="Evaluation of Unico’s parametric statisitcal testing p-values for cell-type level differential methylation with age in whole-blood       datasets (Liu et al.). Presented are quantile-quantile plots with log-transformed expected p-values versus      the observed p-values. Red horizontal dashed lines indicate the Bonferroni-corrected threshold, adjusting for the number of CpGs and        the number of cell-types under test." width="675px"><p class="caption">
Evaluation of Unico’s parametric statisitcal testing p-values for
cell-type level differential methylation with age in whole-blood
datasets (Liu et al.). Presented are quantile-quantile plots with
log-transformed expected p-values versus the observed p-values. Red
horizontal dashed lines indicate the Bonferroni-corrected threshold,
adjusting for the number of CpGs and the number of cell-types under
test.
</p>
</div>
<p>We now check the reported joint association testing results:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">liu.joint.pvals</span>    <span class="op">=</span> <span class="va">unico.liu</span><span class="op">$</span><span class="va">params.hat</span><span class="op">$</span><span class="va">parametric</span><span class="op">$</span><span class="va">gammas_hat_pvals.joint</span><span class="op">[</span>, <span class="st">"age"</span><span class="op">]</span></span>
<span><span class="va">hannum.joint.pvals</span> <span class="op">=</span> <span class="va">unico.hannum</span><span class="op">$</span><span class="va">params.hat</span><span class="op">$</span><span class="va">parametric</span><span class="op">$</span><span class="va">gammas_hat_pvals.joint</span><span class="op">[</span>, <span class="st">"age"</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">liu.joint.pvals</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">/</span><span class="va">m</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">hannum.joint.pvals</span><span class="op">[</span><span class="va">liu.joint.pvals</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">/</span><span class="va">m</span><span class="op">]</span> <span class="op">&lt;</span> <span class="op">(</span><span class="fl">0.05</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">liu.joint.pvals</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">/</span><span class="va">m</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Using Liu et al.’s data as discovery, and Hannum et al.’s data as
validation set. We have 159 out of 206 discovery replicated.</p>
<div class="section level3">
<h3 id="perform-association-testing-under-no-distribution-assumption">Perform association testing under no distribution assumption<a class="anchor" aria-label="anchor" href="#perform-association-testing-under-no-distribution-assumption"></a>
</h3>
<p>Alternatively, we can ask Unico to perform association testing under
no distribution assumption.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">unico.liu</span><span class="op">$</span><span class="va">params.hat</span> <span class="op">=</span> <span class="fu"><a href="../reference/association_asymptotic.html">association_asymptotic</a></span><span class="op">(</span>X <span class="op">=</span> <span class="va">liu</span><span class="op">$</span><span class="va">X</span>, <span class="va">unico.liu</span><span class="op">$</span><span class="va">params.hat</span><span class="op">)</span></span>
<span><span class="va">liu.marg.pvals.asym</span>  <span class="op">=</span> <span class="va">unico.liu</span><span class="op">$</span><span class="va">params.hat</span><span class="op">$</span><span class="va">asymptotic</span><span class="op">$</span><span class="va">gammas_hat_pvals</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">source.ids</span>, <span class="st">".age"</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<p>Empirically, we observed that asymptotically-derived p-values are
highly correlated with their parametric counterparts</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">qq_compare_g</span> <span class="op">=</span> <span class="fu">plot_qq_compare</span><span class="op">(</span>pvals_mat1 <span class="op">=</span> <span class="va">liu.marg.pvals</span>,</span>
<span>                               pvals_mat2 <span class="op">=</span> <span class="va">liu.marg.pvals.asym</span>,</span>
<span>                               labels <span class="op">=</span> <span class="va">source.ids</span>, </span>
<span>                               ggarrange.nrow <span class="op">=</span> <span class="fl">1</span>, ggarrange.ncol <span class="op">=</span> <span class="va">k</span>, </span>
<span>                               alpha <span class="op">=</span> <span class="fl">0.05</span>, text.size <span class="op">=</span> <span class="fl">20</span>,</span>
<span>                               xlab <span class="op">=</span> <span class="st">"Parametric"</span>, ylab <span class="op">=</span> <span class="st">"Asymptotic"</span>, </span>
<span>                               title <span class="op">=</span> <span class="st">"Parametric vs Asymptotic pvals"</span><span class="op">)</span></span>
<span><span class="va">qq_compare_g</span></span></code></pre></div>
<div class="figure">
<img src="qq_compare.png" alt="Evaluation of Unico’s asymptotically-derived p-values under non-parametric testing       for cell-type level differential methylation with age in whole-blood dataset (Liu et al.). Presented are        scatter plots showing log-transformed p-values under the assumption that methylation levels     are normally distributed (“Parametric”) versus the corresponding log-transformed p-values of        a non-parametric test (“Asymptotic”)." width="675px"><p class="caption">
Evaluation of Unico’s asymptotically-derived p-values under
non-parametric testing for cell-type level differential methylation with
age in whole-blood dataset (Liu et al.). Presented are scatter plots
showing log-transformed p-values under the assumption that methylation
levels are normally distributed (“Parametric”) versus the corresponding
log-transformed p-values of a non-parametric test (“Asymptotic”).
</p>
</div>
</div>
<div class="section level3">
<h3 id="calibration-of-asymptotically-derived-p-values-under-null">Calibration of asymptotically-derived p-values under null<a class="anchor" aria-label="anchor" href="#calibration-of-asymptotically-derived-p-values-under-null"></a>
</h3>
<p>To evaluate whether Unico’s asymptotically-derived p-values are well
calibrated under the null, we can shuffle the covariates of interest,
refit the model and ask for cell-type level association in this permuted
data set. We will use age as our running example. Since we now
arbitrarily assigned age to each sample, irrespective to his or her true
biological age, it is expected that his or her methylation profile
reflects no signal with respect to our assigned age, and thus we should
obtain close to 0 detection.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">C1.shuffle</span> <span class="op">=</span> <span class="va">hannum</span><span class="op">$</span><span class="va">cov</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"age"</span>, <span class="st">"sex"</span>, <span class="st">"ethnicity"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">C1.shuffle</span><span class="op">[</span>, <span class="st">"age"</span><span class="op">]</span> <span class="op">=</span> <span class="va">hannum</span><span class="op">$</span><span class="va">cov</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">hannum</span><span class="op">$</span><span class="va">cov</span><span class="op">)</span><span class="op">)</span>, <span class="st">"age"</span><span class="op">]</span></span>
<span></span>
<span><span class="va">unico.hannum.shuffle</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">unico.hannum.shuffle</span><span class="op">$</span><span class="va">params.hat</span> <span class="op">=</span> <span class="fu"><a href="../reference/Unico.html">Unico</a></span><span class="op">(</span>X  <span class="op">=</span> <span class="va">hannum</span><span class="op">$</span><span class="va">X</span>, W <span class="op">=</span> <span class="va">hannum</span><span class="op">$</span><span class="va">W</span>,</span>
<span>                                        C1 <span class="op">=</span> <span class="va">C1.shuffle</span>,</span>
<span>                                        C2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">hannum</span><span class="op">$</span><span class="va">ctrl_pcs</span>, <span class="va">hannum</span><span class="op">$</span><span class="va">cov</span><span class="op">[</span>,<span class="st">"plate"</span>, drop <span class="op">=</span> <span class="cn">F</span><span class="op">]</span><span class="op">)</span><span class="op">)</span> </span>
<span><span class="va">unico.hannum.shuffle</span><span class="op">$</span><span class="va">params.hat</span> <span class="op">=</span> <span class="fu"><a href="../reference/association_asymptotic.html">association_asymptotic</a></span><span class="op">(</span>X <span class="op">=</span> <span class="va">hannum</span><span class="op">$</span><span class="va">X</span>, <span class="va">unico.hannum.shuffle</span><span class="op">$</span><span class="va">params.hat</span><span class="op">)</span></span>
<span><span class="va">unico.marg.pvals.asym.calib</span> <span class="op">=</span> <span class="va">unico.hannum.shuffle</span><span class="op">$</span><span class="va">params.hat</span><span class="op">$</span><span class="va">asymptotic</span><span class="op">$</span><span class="va">gammas_hat_pvals</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">source.ids</span>, <span class="st">".age"</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<p>Indeed, under the null, we see the quantile-quantile plots fall along
the diagonal</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">qq_calib_g</span> <span class="op">=</span> <span class="fu">plot_qq</span><span class="op">(</span>pvals_mat <span class="op">=</span> <span class="va">unico.marg.pvals.asym.calib</span>,</span>
<span>                     labels <span class="op">=</span> <span class="va">source.ids</span>, </span>
<span>                     ggarrange.nrow <span class="op">=</span> <span class="fl">1</span>, ggarrange.ncol <span class="op">=</span> <span class="va">k</span>, </span>
<span>                     alpha <span class="op">=</span> <span class="fl">0.5</span>, text.size <span class="op">=</span> <span class="fl">20</span>,</span>
<span>                     title <span class="op">=</span> <span class="st">"Calibration of asymptotic association testing"</span><span class="op">)</span></span>
<span><span class="va">qq_calib_g</span></span></code></pre></div>
<div class="figure">
<img src="qq_calib.png" alt="Evaluation of the null distribution of Unico’s asymptotically-derived p-values under       non-parametric testing for cell-type level differential methylation with age in whole-blood     datasets (Hannum et al.). Presented are quantile-quantile plots with log-transformed expected p-values versus       the observed p-values under permutations of the condition (i.e., under the null). Red horizontal        dashed lines indicate the Bonferroni-corrected threshold, adjusting for the number of CpGs and      the number of cell-types under test." width="675px"><p class="caption">
Evaluation of the null distribution of Unico’s asymptotically-derived
p-values under non-parametric testing for cell-type level differential
methylation with age in whole-blood datasets (Hannum et al.). Presented
are quantile-quantile plots with log-transformed expected p-values
versus the observed p-values under permutations of the condition (i.e.,
under the null). Red horizontal dashed lines indicate the
Bonferroni-corrected threshold, adjusting for the number of CpGs and the
number of cell-types under test.
</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Zeyuan Chen, Elior Rahmani.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
